// three-stdlib/renderers/nodes/accessors/CameraNode.js
"use strict";
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
const Node = require("../core/Node.js");
const Vector3Node = require("../inputs/Vector3Node.js");
const Matrix4Node = require("../inputs/Matrix4Node.js");
const constants = require("../core/constants.js");
const _CameraNode = class extends Node {
  constructor(scope = _CameraNode.POSITION) {
    super();
    this.updateType = constants.NodeUpdateType.Frame;
    this.scope = scope;
    this._inputNode = null;
  }
  getType() {
    const scope = this.scope;
    if (scope === _CameraNode.PROJECTION || scope === _CameraNode.VIEW) {
      return "mat4";
    }
    return "vec3";
  }
  update(frame) {
    const camera = frame.camera;
    const inputNode = this._inputNode;
    const scope = this.scope;
    if (scope === _CameraNode.PROJECTION) {
      inputNode.value = camera.projectionMatrix;
    } else if (scope === _CameraNode.VIEW) {
      inputNode.value = camera.matrixWorldInverse;
    } else if (scope === _CameraNode.POSITION) {
      camera.getWorldPosition(inputNode.value);
    }
  }
  generate(builder, output) {
    const nodeData = builder.getDataFromNode(this);
    let inputNode = this._inputNode;
    if (nodeData.inputNode === void 0) {
      const scope = this.scope;
      if (scope === _CameraNode.PROJECTION || scope === _CameraNode.VIEW) {
        if (inputNode === null || inputNode.isMatrix4Node !== true) {
          inputNode = new Matrix4Node(null);
        }
      } else if (scope === _CameraNode.POSITION) {
        if (inputNode === null || inputNode.isVector3Node !== true) {
          inputNode = new Vector3Node();
        }
      }
      this._inputNode = inputNode;
      nodeData.inputNode = inputNode;
    }
    return inputNode.build(builder, output);
  }
};
let CameraNode = _CameraNode;
__publicField(CameraNode, "POSITION", "position");
__publicField(CameraNode, "PROJECTION", "projection");
__publicField(CameraNode, "VIEW", "view");
module.exports = CameraNode;
