// three-stdlib/nodes/core/InputNode.mjs
import Node from "./Node.mjs";
import { getValueType, getValueFromType } from "./NodeUtils.mjs";
class InputNode extends Node {
  constructor(value, nodeType = null) {
    super(nodeType);
    this.value = value;
  }
  getNodeType() {
    if (this.nodeType === null) {
      return getValueType(this.value);
    }
    return this.nodeType;
  }
  getInputType(builder) {
    return this.getNodeType(builder);
  }
  serialize(data) {
    var _a, _b;
    super.serialize(data);
    data.value = ((_b = (_a = this.value) == null ? void 0 : _a.toArray) == null ? void 0 : _b.call(_a)) || this.value;
    data.valueType = getValueType(this.value);
    data.nodeType = this.nodeType;
  }
  deserialize(data) {
    var _a, _b;
    super.deserialize(data);
    this.nodeType = data.nodeType;
    this.value = getValueFromType(data.valueType);
    this.value = ((_b = (_a = this.value) == null ? void 0 : _a.fromArray) == null ? void 0 : _b.call(_a, data.value)) || data.value;
  }
  generate() {
    console.warn("Abstract function.");
  }
}
InputNode.prototype.isInputNode = true;
export {
  InputNode as default
};
