// three-stdlib/index.mjs
import "./offscreen/scene.mjs";
import "./offscreen/offscreen.mjs";
import "./offscreen/jank.mjs";
import { MD2CharacterComplex } from "./misc/MD2CharacterComplex.mjs";
import { ConvexObjectBreaker } from "./misc/ConvexObjectBreaker.mjs";
import { MorphBlendMesh } from "./misc/MorphBlendMesh.mjs";
import { GPUComputationRenderer } from "./misc/GPUComputationRenderer.mjs";
import { Gyroscope } from "./misc/Gyroscope.mjs";
import { MorphAnimMesh } from "./misc/MorphAnimMesh.mjs";
import { RollerCoasterGeometry, RollerCoasterLiftersGeometry, RollerCoasterShadowGeometry, SkyGeometry, TreesGeometry } from "./misc/RollerCoaster.mjs";
import { getErrorMessage, getWebGL2ErrorMessage, getWebGLErrorMessage, isWebGL2Available, isWebGLAvailable } from "./misc/WebGL.mjs";
import { MD2Character } from "./misc/MD2Character.mjs";
import { VolumeSlice } from "./misc/VolumeSlice.mjs";
import { TubePainter } from "./misc/TubePainter.mjs";
import { Volume } from "./misc/Volume.mjs";
import { ProgressiveLightMap } from "./misc/ProgressiveLightmap.mjs";
import { CSS2DObject, CSS2DRenderer } from "./renderers/CSS2DRenderer.mjs";
import "./renderers/nodes/core/constants.mjs";
import "./renderers/nodes/core/NodeBuilder.mjs";
import "./renderers/nodes/core/InputNode.mjs";
import "./renderers/nodes/core/NodeUniform.mjs";
import "./renderers/nodes/core/NodeVary.mjs";
import "./renderers/nodes/core/NodeFrame.mjs";
import "./renderers/nodes/core/Node.mjs";
import "./renderers/nodes/core/NodeSlot.mjs";
import "./renderers/nodes/core/NodeAttribute.mjs";
import "./renderers/nodes/core/VaryNode.mjs";
import "./renderers/nodes/core/AttributeNode.mjs";
import "./renderers/nodes/accessors/CameraNode.mjs";
import "./renderers/nodes/accessors/UVNode.mjs";
import "./renderers/nodes/accessors/ModelNode.mjs";
import "./renderers/nodes/accessors/NormalNode.mjs";
import "./renderers/nodes/accessors/PositionNode.mjs";
import "./renderers/nodes/accessors/ModelViewProjectionNode.mjs";
import "./renderers/nodes/utils/SwitchNode.mjs";
import "./renderers/nodes/utils/TimerNode.mjs";
import "./renderers/nodes/math/OperatorNode.mjs";
import "./renderers/nodes/math/MathNode.mjs";
import "./renderers/nodes/inputs/ColorNode.mjs";
import "./renderers/nodes/inputs/Vector3Node.mjs";
import "./renderers/nodes/inputs/Matrix4Node.mjs";
import "./renderers/nodes/inputs/TextureNode.mjs";
import "./renderers/nodes/inputs/FloatNode.mjs";
import "./renderers/nodes/inputs/Vector2Node.mjs";
import "./renderers/nodes/inputs/Vector4Node.mjs";
import "./renderers/nodes/inputs/Matrix3Node.mjs";
import { BlendColorFactor, GPUAddressMode, GPUBindingType, GPUBlendFactor, GPUBlendOperation, GPUColorWriteFlags, GPUCompareFunction, GPUCullMode, GPUFilterMode, GPUFrontFace, GPUIndexFormat, GPUInputStepMode, GPULoadOp, GPUPrimitiveTopology, GPUStencilOperation, GPUStoreOp, GPUTextureDimension, GPUTextureFormat, GPUTextureViewDimension, GPUVertexFormat, OneMinusBlendColorFactor } from "./renderers/webgpu/constants.mjs";
import "./renderers/webgpu/WebGPUTextures.mjs";
import "./renderers/webgpu/WebGPUInfo.mjs";
import "./renderers/webgpu/WebGPURenderLists.mjs";
import "./renderers/webgpu/WebGPUTextureRenderer.mjs";
import "./renderers/webgpu/WebGPURenderPipelines.mjs";
import "./renderers/webgpu/WebGPUGeometries.mjs";
import "./renderers/webgpu/nodes/WebGPUNodeBuilder.mjs";
import { ColorNodeUniform, FloatNodeUniform, Matrix3NodeUniform, Matrix4NodeUniform, Vector2NodeUniform, Vector3NodeUniform, Vector4NodeUniform } from "./renderers/webgpu/nodes/WebGPUNodeUniform.mjs";
import "./renderers/webgpu/nodes/WebGPUNodeUniformsGroup.mjs";
import "./renderers/webgpu/nodes/ShaderLib.mjs";
import "./renderers/webgpu/nodes/WebGPUNodes.mjs";
import "./renderers/webgpu/WebGPUBindings.mjs";
import { WebGPUSampled3DTexture, WebGPUSampledArrayTexture, WebGPUSampledCubeTexture, WebGPUSampledTexture } from "./renderers/webgpu/WebGPUSampledTexture.mjs";
import "./renderers/webgpu/WebGPU.mjs";
import "./renderers/webgpu/WebGPUUniformsGroup.mjs";
import "./renderers/webgpu/WebGPURenderer.mjs";
import "./renderers/webgpu/WebGPUComputePipelines.mjs";
import { ColorUniform, FloatUniform, Matrix3Uniform, Matrix4Uniform, Vector2Uniform, Vector3Uniform, Vector4Uniform } from "./renderers/webgpu/WebGPUUniform.mjs";
import "./renderers/webgpu/WebGPUObjects.mjs";
import "./renderers/webgpu/WebGPUSampler.mjs";
import "./renderers/webgpu/WebGPUBackground.mjs";
import "./renderers/webgpu/WebGPUBinding.mjs";
import "./renderers/webgpu/WebGPUProperties.mjs";
import "./renderers/webgpu/WebGPUAttributes.mjs";
import "./renderers/webgpu/WebGPUStorageBuffer.mjs";
import "./renderers/webgpu/WebGPUTextureUtils.mjs";
import { CSS3DObject, CSS3DRenderer, CSS3DSprite } from "./renderers/CSS3DRenderer.mjs";
import { Projector, RenderableFace, RenderableLine, RenderableObject, RenderableSprite, RenderableVertex } from "./renderers/Projector.mjs";
import { SVGObject, SVGRenderer } from "./renderers/SVGRenderer.mjs";
import { FlakesTexture } from "./textures/FlakesTexture.mjs";
import { Flow, InstancedFlow, getUniforms, initSplineTexture, modifyShader, updateSplineTexture } from "./modifiers/CurveModifier.mjs";
import { SimplifyModifier } from "./modifiers/SimplifyModifier.mjs";
import { EdgeSplitModifier } from "./modifiers/EdgeSplitModifier.mjs";
import { TessellateModifier } from "./modifiers/TessellateModifier.mjs";
import { fromType } from "./nodes/Nodes.mjs";
import { GLTFExporter } from "./exporters/GLTFExporter.mjs";
import { USDZExporter } from "./exporters/USDZExporter.mjs";
import { PLYExporter } from "./exporters/PLYExporter.mjs";
import { DRACOExporter } from "./exporters/DRACOExporter.mjs";
import { ColladaExporter } from "./exporters/ColladaExporter.mjs";
import { MMDExporter } from "./exporters/MMDExporter.mjs";
import { STLExporter } from "./exporters/STLExporter.mjs";
import { OBJExporter } from "./exporters/OBJExporter.mjs";
import { RoomEnvironment } from "./environments/RoomEnvironment.mjs";
import { AnimationClipCreator } from "./animation/AnimationClipCreator.mjs";
import { CCDIKHelper, CCDIKSolver } from "./animation/CCDIKSolver.mjs";
import { MMDPhysics } from "./animation/MMDPhysics.mjs";
import { MMDAnimationHelper } from "./animation/MMDAnimationHelper.mjs";
import { Reflector } from "./objects/Reflector.mjs";
import { Refractor } from "./objects/Refractor.mjs";
import { ShadowMesh } from "./objects/ShadowMesh.mjs";
import { Lensflare, LensflareElement } from "./objects/Lensflare.mjs";
import { Water } from "./objects/Water.mjs";
import { MarchingCubes, edgeTable, triTable } from "./objects/MarchingCubes.mjs";
import { LightningStorm } from "./objects/LightningStorm.mjs";
import { ReflectorRTT } from "./objects/ReflectorRTT.mjs";
import { ReflectorForSSRPass } from "./objects/ReflectorForSSRPass.mjs";
import { Sky } from "./objects/Sky.mjs";
import { Water2 } from "./objects/Water2.mjs";
import { GroundProjectedEnv } from "./objects/GroundProjectedEnv.mjs";
import { SceneUtils } from "./utils/SceneUtils.mjs";
import { UVsDebug } from "./utils/UVsDebug.mjs";
import { GeometryUtils } from "./utils/GeometryUtils.mjs";
import { RoughnessMipmapper } from "./utils/RoughnessMipmapper.mjs";
import { SkeletonUtils } from "./utils/SkeletonUtils.mjs";
import { ShadowMapViewer } from "./utils/ShadowMapViewer.mjs";
import { computeMorphedAttributes, estimateBytesUsed, interleaveAttributes, mergeBufferAttributes, mergeBufferGeometries, mergeVertices, toTrianglesDrawMode } from "./utils/BufferGeometryUtils.mjs";
import { GeometryCompressionUtils, PackedPhongMaterial } from "./utils/GeometryCompressionUtils.mjs";
import { CinematicCamera } from "./cameras/CinematicCamera.mjs";
import { ConvexHull } from "./math/ConvexHull.mjs";
import { MeshSurfaceSampler } from "./math/MeshSurfaceSampler.mjs";
import { SimplexNoise } from "./math/SimplexNoise.mjs";
import { OBB } from "./math/OBB.mjs";
import { Capsule } from "./math/Capsule.mjs";
import { ColorConverter } from "./math/ColorConverter.mjs";
import { ImprovedNoise } from "./math/ImprovedNoise.mjs";
import { Octree } from "./math/Octree.mjs";
import { ColorMapKeywords, Lut } from "./math/Lut.mjs";
import { CameraControls, MapControlsExp, OrbitControlsExp, STATE, TrackballControlsExp } from "./controls/experimental/CameraControls.mjs";
import { FirstPersonControls } from "./controls/FirstPersonControls.mjs";
import { TransformControls, TransformControlsGizmo, TransformControlsPlane } from "./controls/TransformControls.mjs";
import { DragControls } from "./controls/DragControls.mjs";
import { PointerLockControls } from "./controls/PointerLockControls.mjs";
import { DeviceOrientationControls } from "./controls/DeviceOrientationControls.mjs";
import { TrackballControls } from "./controls/TrackballControls.mjs";
import { MapControls, OrbitControls } from "./controls/OrbitControls.mjs";
import { ArcballControls } from "./controls/ArcballControls.mjs";
import { FlyControls } from "./controls/FlyControls.mjs";
import { LUTPass } from "./postprocessing/LUTPass.mjs";
import { ClearPass } from "./postprocessing/ClearPass.mjs";
import { GlitchPass } from "./postprocessing/GlitchPass.mjs";
import { HalftonePass } from "./postprocessing/HalftonePass.mjs";
import { SMAAPass } from "./postprocessing/SMAAPass.mjs";
import { FilmPass } from "./postprocessing/FilmPass.mjs";
import { OutlinePass } from "./postprocessing/OutlinePass.mjs";
import { SSAOPass } from "./postprocessing/SSAOPass.mjs";
import { SavePass } from "./postprocessing/SavePass.mjs";
import { BokehPass } from "./postprocessing/BokehPass.mjs";
import { FullScreenQuad, Pass } from "./postprocessing/Pass.mjs";
import { TexturePass } from "./postprocessing/TexturePass.mjs";
import { AdaptiveToneMappingPass } from "./postprocessing/AdaptiveToneMappingPass.mjs";
import { UnrealBloomPass } from "./postprocessing/UnrealBloomPass.mjs";
import { CubeTexturePass } from "./postprocessing/CubeTexturePass.mjs";
import { SAOPass } from "./postprocessing/SAOPass.mjs";
import { AfterimagePass } from "./postprocessing/AfterimagePass.mjs";
import { ClearMaskPass, MaskPass } from "./postprocessing/MaskPass.mjs";
import { EffectComposer } from "./postprocessing/EffectComposer.mjs";
import { DotScreenPass } from "./postprocessing/DotScreenPass.mjs";
import { SSRPass } from "./postprocessing/SSRPass.mjs";
import { TAARenderPass } from "./postprocessing/TAARenderPass.mjs";
import { ShaderPass } from "./postprocessing/ShaderPass.mjs";
import { SSAARenderPass } from "./postprocessing/SSAARenderPass.mjs";
import { RenderPass } from "./postprocessing/RenderPass.mjs";
import { RenderPixelatedPass } from "./postprocessing/RenderPixelatedPass.mjs";
import { BloomPass } from "./postprocessing/BloomPass.mjs";
import { ARButton } from "./webxr/ARButton.mjs";
import { OculusHandModel } from "./webxr/OculusHandModel.mjs";
import { OculusHandPointerModel } from "./webxr/OculusHandPointerModel.mjs";
import { createText } from "./webxr/Text2D.mjs";
import { VRButton } from "./webxr/VRButton.mjs";
import { XRControllerModelFactory } from "./webxr/XRControllerModelFactory.mjs";
import { XREstimatedLight } from "./webxr/XREstimatedLight.mjs";
import { XRHandMeshModel } from "./webxr/XRHandMeshModel.mjs";
import { XRHandModelFactory } from "./webxr/XRHandModelFactory.mjs";
import { XRHandPrimitiveModel } from "./webxr/XRHandPrimitiveModel.mjs";
import { ParametricGeometries } from "./geometries/ParametricGeometries.mjs";
import { ParametricGeometry } from "./geometries/ParametricGeometry.mjs";
import { ConvexGeometry } from "./geometries/ConvexGeometry.mjs";
import { LightningStrike } from "./geometries/LightningStrike.mjs";
import { RoundedBoxGeometry } from "./geometries/RoundedBoxGeometry.mjs";
import { BoxLineGeometry } from "./geometries/BoxLineGeometry.mjs";
import { DecalGeometry, DecalVertex } from "./geometries/DecalGeometry.mjs";
import { TeapotGeometry } from "./geometries/TeapotGeometry.mjs";
import { TextGeometry, TextGeometry as TextGeometry2 } from "./geometries/TextGeometry.mjs";
import { CSM } from "./csm/CSM.mjs";
import { CSMFrustum } from "./csm/CSMFrustum.mjs";
import { CSMHelper } from "./csm/CSMHelper.mjs";
import { CSMShader } from "./csm/CSMShader.mjs";
import "./shaders/types.mjs";
import { ACESFilmicToneMappingShader } from "./shaders/ACESFilmicToneMappingShader.mjs";
import { AfterimageShader } from "./shaders/AfterimageShader.mjs";
import { BasicShader } from "./shaders/BasicShader.mjs";
import { BleachBypassShader } from "./shaders/BleachBypassShader.mjs";
import { BlendShader } from "./shaders/BlendShader.mjs";
import { BokehShader } from "./shaders/BokehShader.mjs";
import { BokehShader2 } from "./shaders/BokehShader2.mjs";
import { BrightnessContrastShader } from "./shaders/BrightnessContrastShader.mjs";
import { ColorCorrectionShader } from "./shaders/ColorCorrectionShader.mjs";
import { ColorifyShader } from "./shaders/ColorifyShader.mjs";
import { ConvolutionShader } from "./shaders/ConvolutionShader.mjs";
import { CopyShader } from "./shaders/CopyShader.mjs";
import { DOFMipMapShader } from "./shaders/DOFMipMapShader.mjs";
import { BlurShaderUtils, DepthLimitedBlurShader } from "./shaders/DepthLimitedBlurShader.mjs";
import { DigitalGlitch } from "./shaders/DigitalGlitch.mjs";
import { DotScreenShader } from "./shaders/DotScreenShader.mjs";
import { FXAAShader } from "./shaders/FXAAShader.mjs";
import { FilmShader } from "./shaders/FilmShader.mjs";
import { FocusShader } from "./shaders/FocusShader.mjs";
import { FreiChenShader } from "./shaders/FreiChenShader.mjs";
import { FresnelShader } from "./shaders/FresnelShader.mjs";
import { GammaCorrectionShader } from "./shaders/GammaCorrectionShader.mjs";
import { GodRaysCombineShader, GodRaysDepthMaskShader, GodRaysFakeSunShader, GodRaysGenerateShader } from "./shaders/GodRaysShader.mjs";
import { HalftoneShader } from "./shaders/HalftoneShader.mjs";
import { HorizontalBlurShader } from "./shaders/HorizontalBlurShader.mjs";
import { HorizontalTiltShiftShader } from "./shaders/HorizontalTiltShiftShader.mjs";
import { HueSaturationShader } from "./shaders/HueSaturationShader.mjs";
import { KaleidoShader } from "./shaders/KaleidoShader.mjs";
import { LuminosityHighPassShader } from "./shaders/LuminosityHighPassShader.mjs";
import { LuminosityShader } from "./shaders/LuminosityShader.mjs";
import { MirrorShader } from "./shaders/MirrorShader.mjs";
import { NormalMapShader } from "./shaders/NormalMapShader.mjs";
import { ParallaxShader } from "./shaders/ParallaxShader.mjs";
import { PixelShader } from "./shaders/PixelShader.mjs";
import { RGBShiftShader } from "./shaders/RGBShiftShader.mjs";
import { SAOShader } from "./shaders/SAOShader.mjs";
import { SMAABlendShader, SMAAEdgesShader, SMAAWeightsShader } from "./shaders/SMAAShader.mjs";
import { SSAOBlurShader, SSAODepthShader, SSAOShader } from "./shaders/SSAOShader.mjs";
import { SSRBlurShader, SSRDepthShader, SSRShader } from "./shaders/SSRShader.mjs";
import { SepiaShader } from "./shaders/SepiaShader.mjs";
import { SobelOperatorShader } from "./shaders/SobelOperatorShader.mjs";
import { SubsurfaceScatteringShader } from "./shaders/SubsurfaceScatteringShader.mjs";
import { TechnicolorShader } from "./shaders/TechnicolorShader.mjs";
import { ToneMapShader } from "./shaders/ToneMapShader.mjs";
import { ToonShader1, ToonShader2, ToonShaderDotted, ToonShaderHatching } from "./shaders/ToonShader.mjs";
import { TriangleBlurShader } from "./shaders/TriangleBlurShader.mjs";
import { UnpackDepthRGBAShader } from "./shaders/UnpackDepthRGBAShader.mjs";
import { VerticalBlurShader } from "./shaders/VerticalBlurShader.mjs";
import { VerticalTiltShiftShader } from "./shaders/VerticalTiltShiftShader.mjs";
import { VignetteShader } from "./shaders/VignetteShader.mjs";
import { VolumeRenderShader1 } from "./shaders/VolumeShader.mjs";
import { WaterRefractionShader } from "./shaders/WaterRefractionShader.mjs";
import { SelectionHelper } from "./interactive/SelectionHelper.mjs";
import { SelectionBox } from "./interactive/SelectionBox.mjs";
import { AmmoPhysics } from "./physics/AmmoPhysics.mjs";
import { ParallaxBarrierEffect } from "./effects/ParallaxBarrierEffect.mjs";
import { PeppersGhostEffect } from "./effects/PeppersGhostEffect.mjs";
import { OutlineEffect } from "./effects/OutlineEffect.mjs";
import { AnaglyphEffect } from "./effects/AnaglyphEffect.mjs";
import { AsciiEffect } from "./effects/AsciiEffect.mjs";
import { StereoEffect } from "./effects/StereoEffect.mjs";
import { FBXLoader } from "./loaders/FBXLoader.mjs";
import { Font, FontLoader } from "./loaders/FontLoader.mjs";
import { TGALoader } from "./loaders/TGALoader.mjs";
import { LUTCubeLoader } from "./loaders/LUTCubeLoader.mjs";
import { NRRDLoader } from "./loaders/NRRDLoader.mjs";
import { STLLoader } from "./loaders/STLLoader.mjs";
import { MTLLoader } from "./loaders/MTLLoader.mjs";
import { XLoader } from "./loaders/XLoader.mjs";
import { BVHLoader } from "./loaders/BVHLoader.mjs";
import { KMZLoader } from "./loaders/KMZLoader.mjs";
import { VRMLoader } from "./loaders/VRMLoader.mjs";
import { VRMLLoader } from "./loaders/VRMLLoader.mjs";
import { KTX2Loader } from "./loaders/KTX2Loader.mjs";
import { LottieLoader } from "./loaders/LottieLoader.mjs";
import { TTFLoader } from "./loaders/TTFLoader.mjs";
import { RGBELoader } from "./loaders/RGBELoader.mjs";
import { AssimpLoader } from "./loaders/AssimpLoader.mjs";
import { ColladaLoader } from "./loaders/ColladaLoader.mjs";
import { MDDLoader } from "./loaders/MDDLoader.mjs";
import { EXRLoader } from "./loaders/EXRLoader.mjs";
import { ThreeMFLoader } from "./loaders/3MFLoader.mjs";
import { XYZLoader } from "./loaders/XYZLoader.mjs";
import { VTKLoader } from "./loaders/VTKLoader.mjs";
import { LUT3dlLoader } from "./loaders/LUT3dlLoader.mjs";
import { DDSLoader } from "./loaders/DDSLoader.mjs";
import { PVRLoader } from "./loaders/PVRLoader.mjs";
import { GCodeLoader } from "./loaders/GCodeLoader.mjs";
import { BasisTextureLoader } from "./loaders/BasisTextureLoader.mjs";
import { TDSLoader } from "./loaders/TDSLoader.mjs";
import { LDrawLoader } from "./loaders/LDrawLoader.mjs";
import { GLTFLoader } from "./loaders/GLTFLoader.mjs";
import { LWO3Parser } from "./loaders/lwo/LWO3Parser.mjs";
import { LWO2Parser } from "./loaders/lwo/LWO2Parser.mjs";
import { IFFParser } from "./loaders/lwo/IFFParser.mjs";
import { SVGLoader } from "./loaders/SVGLoader.mjs";
import { Rhino3dmLoader } from "./loaders/3DMLoader.mjs";
import { OBJLoader } from "./loaders/OBJLoader.mjs";
import { AMFLoader } from "./loaders/AMFLoader.mjs";
import { MMDLoader } from "./loaders/MMDLoader.mjs";
import { MD2Loader } from "./loaders/MD2Loader.mjs";
import { KTXLoader } from "./loaders/KTXLoader.mjs";
import { TiltLoader } from "./loaders/TiltLoader.mjs";
import { DRACOLoader } from "./loaders/DRACOLoader.mjs";
import { HDRCubeTextureLoader } from "./loaders/HDRCubeTextureLoader.mjs";
import { PDBLoader } from "./loaders/PDBLoader.mjs";
import { PRWMLoader } from "./loaders/PRWMLoader.mjs";
import { RGBMLoader } from "./loaders/RGBMLoader.mjs";
import { VOXDataTexture3D, VOXLoader, VOXMesh } from "./loaders/VOXLoader.mjs";
import { NodeMaterialLoaderUtils } from "./loaders/NodeMaterialLoader.mjs";
import { PCDLoader } from "./loaders/PCDLoader.mjs";
import { LWOLoader } from "./loaders/LWOLoader.mjs";
import { PLYLoader } from "./loaders/PLYLoader.mjs";
import { LineSegmentsGeometry } from "./lines/LineSegmentsGeometry.mjs";
import { LineGeometry } from "./lines/LineGeometry.mjs";
import { Wireframe } from "./lines/Wireframe.mjs";
import { WireframeGeometry2 } from "./lines/WireframeGeometry2.mjs";
import { Line2 } from "./lines/Line2.mjs";
import { LineMaterial } from "./lines/LineMaterial.mjs";
import { LineSegments2 } from "./lines/LineSegments2.mjs";
import { LightProbeHelper } from "./helpers/LightProbeHelper.mjs";
import { VertexTangentsHelper } from "./helpers/VertexTangentsHelper.mjs";
import { PositionalAudioHelper } from "./helpers/PositionalAudioHelper.mjs";
import { VertexNormalsHelper } from "./helpers/VertexNormalsHelper.mjs";
import { RectAreaLightHelper } from "./helpers/RectAreaLightHelper.mjs";
import { RectAreaLightUniformsLib } from "./lights/RectAreaLightUniformsLib.mjs";
import { LightProbeGenerator } from "./lights/LightProbeGenerator.mjs";
import { calcBSplineDerivatives, calcBSplinePoint, calcBasisFunctionDerivatives, calcBasisFunctions, calcKoverI, calcNURBSDerivatives, calcRationalCurveDerivatives, calcSurfacePoint, findSpan } from "./curves/NURBSUtils.mjs";
import { NURBSCurve } from "./curves/NURBSCurve.mjs";
import { NURBSSurface } from "./curves/NURBSSurface.mjs";
import { CinquefoilKnot, DecoratedTorusKnot4a, DecoratedTorusKnot4b, DecoratedTorusKnot5a, DecoratedTorusKnot5c, FigureEightPolynomialKnot, GrannyKnot, HeartCurve, HelixCurve, KnotCurve, TorusKnot, TrefoilKnot, TrefoilPolynomialKnot, VivianiCurve, scaleTo } from "./curves/CurveExtras.mjs";
import { Face3, Geometry } from "./deprecated/Geometry.mjs";
import { MeshoptDecoder } from "./libs/MeshoptDecoder.mjs";
import { MotionController, MotionControllerConstants, fetchProfile, fetchProfilesList } from "./libs/MotionControllers.mjs";
import { BRDF_GGX, BRDF_Lambert, D_GGX, F_Schlick, PhysicalLightingModel, RE_Direct_Physical, V_GGX_SmithCorrelated, getDistanceAttenuation } from "./nodes/functions/BSDFs.mjs";
import { default as default2 } from "./nodes/materials/LineBasicNodeMaterial.mjs";
import { default as default3 } from "./nodes/materials/MeshBasicNodeMaterial.mjs";
import { default as default4 } from "./nodes/materials/MeshStandardNodeMaterial.mjs";
import { default as default5 } from "./nodes/materials/PointsNodeMaterial.mjs";
import { EPSILON, INFINITY, ShaderNode, abs, acos, add, addTo, alphaTest, and, asin, assign, atan, attribute, bitAnd, bitOr, bitXor, bmat3, bmat4, bool, buffer, bvec2, bvec3, bvec4, cameraPosition, ceil, clamp, color, cond, cos, cross, dFdx, dFdy, degrees, diffuseColor, distance, div, dot, element, equal, exp, exp2, faceforward, float, floor, fract, greaterThan, greaterThanEqual, imat3, imat4, int, inversesqrt, invert, ivec2, ivec3, ivec4, join, label, length, lessThan, lessThanEqual, log, log2, mat3, mat4, max, metalness, min, mix, mod, mul, negate, nodeObject, normalGeometry, normalLocal, normalView, normalWorld, normalize, or, positionLocal, positionView, positionViewDirection, positionWorld, pow, pow2, pow3, pow4, radians, reflect, refract, remainder, roughness, round, sampler, saturate, shiftLeft, shiftRight, sign, sin, smoothstep, specularColor, sqrt, step, sub, tan, temp, texture, transformDirection, transformedNormalView, uint, umat3, umat4, uniform, uv, uvec2, uvec3, uvec4, vec2, vec3, vec4, viewMatrix, xor } from "./nodes/ShaderNode.mjs";
import { default as default6 } from "./nodes/core/ArrayUniformNode.mjs";
import { default as default7 } from "./nodes/core/AttributeNode.mjs";
import { default as default8 } from "./nodes/core/BypassNode.mjs";
import { default as default9 } from "./nodes/core/CodeNode.mjs";
import { default as default10 } from "./nodes/core/ContextNode.mjs";
import { default as default11 } from "./nodes/core/ConstNode.mjs";
import { default as default12 } from "./nodes/core/ExpressionNode.mjs";
import { default as default13 } from "./nodes/core/FunctionCallNode.mjs";
import { default as default14 } from "./nodes/core/FunctionNode.mjs";
import { default as default15 } from "./nodes/core/Node.mjs";
import { default as default16 } from "./nodes/core/NodeAttribute.mjs";
import { default as default17 } from "./nodes/core/NodeBuilder.mjs";
import { default as default18 } from "./nodes/core/NodeCode.mjs";
import { default as default19 } from "./nodes/core/NodeFrame.mjs";
import { default as default20 } from "./nodes/core/NodeFunctionInput.mjs";
import { default as default21 } from "./nodes/core/NodeKeywords.mjs";
import { default as default22 } from "./nodes/core/NodeUniform.mjs";
import { default as default23 } from "./nodes/core/NodeVar.mjs";
import { default as default24 } from "./nodes/core/NodeVary.mjs";
import { default as default25 } from "./nodes/core/PropertyNode.mjs";
import { default as default26 } from "./nodes/core/TempNode.mjs";
import { default as default27 } from "./nodes/core/UniformNode.mjs";
import { default as default28 } from "./nodes/core/VarNode.mjs";
import { default as default29 } from "./nodes/core/VaryNode.mjs";
import { default as default30 } from "./nodes/accessors/BufferNode.mjs";
import { default as default31 } from "./nodes/accessors/CameraNode.mjs";
import { default as default32 } from "./nodes/accessors/CubeTextureNode.mjs";
import { default as default33 } from "./nodes/accessors/MaterialNode.mjs";
import { default as default34 } from "./nodes/accessors/MaterialReferenceNode.mjs";
import { default as default35 } from "./nodes/accessors/ModelNode.mjs";
import { default as default36 } from "./nodes/accessors/ModelViewProjectionNode.mjs";
import { default as default37 } from "./nodes/accessors/NormalNode.mjs";
import { default as default38 } from "./nodes/accessors/Object3DNode.mjs";
import { default as default39 } from "./nodes/accessors/PointUVNode.mjs";
import { default as default40 } from "./nodes/accessors/PositionNode.mjs";
import { default as default41 } from "./nodes/accessors/ReferenceNode.mjs";
import { default as default42 } from "./nodes/accessors/ReflectNode.mjs";
import { default as default43 } from "./nodes/accessors/SkinningNode.mjs";
import { default as default44 } from "./nodes/accessors/TextureNode.mjs";
import { default as default45 } from "./nodes/accessors/UVNode.mjs";
import { default as default46 } from "./nodes/display/ColorSpaceNode.mjs";
import { default as default47 } from "./nodes/display/NormalMapNode.mjs";
import { default as default48 } from "./nodes/math/MathNode.mjs";
import { default as default49 } from "./nodes/math/OperatorNode.mjs";
import { default as default50 } from "./nodes/math/CondNode.mjs";
import { default as default51 } from "./nodes/lights/LightContextNode.mjs";
import { default as default52 } from "./nodes/lights/LightNode.mjs";
import { default as default53 } from "./nodes/lights/LightsNode.mjs";
import { default as default54 } from "./nodes/utils/ArrayElementNode.mjs";
import { default as default55 } from "./nodes/utils/ConvertNode.mjs";
import { default as default56 } from "./nodes/utils/JoinNode.mjs";
import { default as default57 } from "./nodes/utils/SplitNode.mjs";
import { default as default58 } from "./nodes/utils/SpriteSheetUVNode.mjs";
import { default as default59 } from "./nodes/utils/MatcapUVNode.mjs";
import { default as default60 } from "./nodes/utils/OscNode.mjs";
import { default as default61 } from "./nodes/utils/TimerNode.mjs";
import { default as default62 } from "./nodes/procedural/CheckerNode.mjs";
import { default as default63 } from "./nodes/fog/FogNode.mjs";
import { default as default64 } from "./nodes/fog/FogRangeNode.mjs";
import { default as default65 } from "./nodes/loaders/NodeLoader.mjs";
import { default as default66 } from "./nodes/loaders/NodeObjectLoader.mjs";
export {
  ACESFilmicToneMappingShader,
  AMFLoader,
  ARButton,
  AdaptiveToneMappingPass,
  AfterimagePass,
  AfterimageShader,
  AmmoPhysics,
  AnaglyphEffect,
  AnimationClipCreator,
  ArcballControls,
  default54 as ArrayElementNode,
  default6 as ArrayUniformNode,
  AsciiEffect,
  AssimpLoader,
  default7 as AttributeNode,
  BRDF_GGX,
  BRDF_Lambert,
  BVHLoader,
  BasicShader,
  BasisTextureLoader,
  BleachBypassShader,
  BlendColorFactor,
  BlendShader,
  BloomPass,
  BlurShaderUtils,
  BokehPass,
  BokehShader,
  BokehShader2,
  BoxLineGeometry,
  BrightnessContrastShader,
  default30 as BufferNode,
  default8 as BypassNode,
  CCDIKHelper,
  CCDIKSolver,
  CSM,
  CSMFrustum,
  CSMHelper,
  CSMShader,
  CSS2DObject,
  CSS2DRenderer,
  CSS3DObject,
  CSS3DRenderer,
  CSS3DSprite,
  CameraControls,
  default31 as CameraNode,
  Capsule,
  default62 as CheckerNode,
  CinematicCamera,
  CinquefoilKnot,
  ClearMaskPass,
  ClearPass,
  default9 as CodeNode,
  ColladaExporter,
  ColladaLoader,
  ColorConverter,
  ColorCorrectionShader,
  ColorMapKeywords,
  ColorNodeUniform,
  default46 as ColorSpaceNode,
  ColorUniform,
  ColorifyShader,
  default50 as CondNode,
  default11 as ConstNode,
  default10 as ContextNode,
  default55 as ConvertNode,
  ConvexGeometry,
  ConvexHull,
  ConvexObjectBreaker,
  ConvolutionShader,
  CopyShader,
  default32 as CubeTextureNode,
  CubeTexturePass,
  DDSLoader,
  DOFMipMapShader,
  DRACOExporter,
  DRACOLoader,
  D_GGX,
  DecalGeometry,
  DecalVertex,
  DecoratedTorusKnot4a,
  DecoratedTorusKnot4b,
  DecoratedTorusKnot5a,
  DecoratedTorusKnot5c,
  DepthLimitedBlurShader,
  DeviceOrientationControls,
  DigitalGlitch,
  DotScreenPass,
  DotScreenShader,
  DragControls,
  EPSILON,
  EXRLoader,
  EdgeSplitModifier,
  EffectComposer,
  default12 as ExpressionNode,
  FBXLoader,
  FXAAShader,
  F_Schlick,
  Face3,
  FigureEightPolynomialKnot,
  FilmPass,
  FilmShader,
  FirstPersonControls,
  FlakesTexture,
  FloatNodeUniform,
  FloatUniform,
  Flow,
  FlyControls,
  FocusShader,
  default63 as FogNode,
  default64 as FogRangeNode,
  Font,
  FontLoader,
  FreiChenShader,
  FresnelShader,
  FullScreenQuad,
  default13 as FunctionCallNode,
  default14 as FunctionNode,
  GCodeLoader,
  GLTFExporter,
  GLTFLoader,
  GPUAddressMode,
  GPUBindingType,
  GPUBlendFactor,
  GPUBlendOperation,
  GPUColorWriteFlags,
  GPUCompareFunction,
  GPUComputationRenderer,
  GPUCullMode,
  GPUFilterMode,
  GPUFrontFace,
  GPUIndexFormat,
  GPUInputStepMode,
  GPULoadOp,
  GPUPrimitiveTopology,
  GPUStencilOperation,
  GPUStoreOp,
  GPUTextureDimension,
  GPUTextureFormat,
  GPUTextureViewDimension,
  GPUVertexFormat,
  GammaCorrectionShader,
  Geometry,
  GeometryCompressionUtils,
  GeometryUtils,
  GlitchPass,
  GodRaysCombineShader,
  GodRaysDepthMaskShader,
  GodRaysFakeSunShader,
  GodRaysGenerateShader,
  GrannyKnot,
  GroundProjectedEnv,
  Gyroscope,
  HDRCubeTextureLoader,
  HalftonePass,
  HalftoneShader,
  HeartCurve,
  HelixCurve,
  HorizontalBlurShader,
  HorizontalTiltShiftShader,
  HueSaturationShader,
  IFFParser,
  INFINITY,
  ImprovedNoise,
  InstancedFlow,
  default56 as JoinNode,
  KMZLoader,
  KTX2Loader,
  KTXLoader,
  KaleidoShader,
  KnotCurve,
  LDrawLoader,
  LUT3dlLoader,
  LUTCubeLoader,
  LUTPass,
  LWO2Parser,
  LWO3Parser,
  LWOLoader,
  Lensflare,
  LensflareElement,
  default51 as LightContextNode,
  default52 as LightNode,
  LightProbeGenerator,
  LightProbeHelper,
  LightningStorm,
  LightningStrike,
  default53 as LightsNode,
  Line2,
  default2 as LineBasicNodeMaterial,
  LineGeometry,
  LineMaterial,
  LineSegments2,
  LineSegmentsGeometry,
  LottieLoader,
  LuminosityHighPassShader,
  LuminosityShader,
  Lut,
  MD2Character,
  MD2CharacterComplex,
  MD2Loader,
  MDDLoader,
  MMDAnimationHelper,
  MMDExporter,
  MMDLoader,
  MMDPhysics,
  MTLLoader,
  MapControls,
  MapControlsExp,
  MarchingCubes,
  MaskPass,
  default59 as MatcapUVNode,
  default33 as MaterialNode,
  default34 as MaterialReferenceNode,
  default48 as MathNode,
  Matrix3NodeUniform,
  Matrix3Uniform,
  Matrix4NodeUniform,
  Matrix4Uniform,
  default3 as MeshBasicNodeMaterial,
  default4 as MeshStandardNodeMaterial,
  MeshSurfaceSampler,
  MeshoptDecoder,
  MirrorShader,
  default35 as ModelNode,
  default36 as ModelViewProjectionNode,
  MorphAnimMesh,
  MorphBlendMesh,
  MotionController,
  MotionControllerConstants,
  NRRDLoader,
  NURBSCurve,
  NURBSSurface,
  default15 as Node,
  default16 as NodeAttribute,
  default17 as NodeBuilder,
  default18 as NodeCode,
  default19 as NodeFrame,
  default20 as NodeFunctionInput,
  default21 as NodeKeywords,
  default65 as NodeLoader,
  NodeMaterialLoaderUtils,
  default66 as NodeObjectLoader,
  default22 as NodeUniform,
  default23 as NodeVar,
  default24 as NodeVary,
  default47 as NormalMapNode,
  NormalMapShader,
  default37 as NormalNode,
  OBB,
  OBJExporter,
  OBJLoader,
  default38 as Object3DNode,
  Octree,
  OculusHandModel,
  OculusHandPointerModel,
  OneMinusBlendColorFactor,
  default49 as OperatorNode,
  OrbitControls,
  OrbitControlsExp,
  default60 as OscNode,
  OutlineEffect,
  OutlinePass,
  PCDLoader,
  PDBLoader,
  PLYExporter,
  PLYLoader,
  PRWMLoader,
  PVRLoader,
  PackedPhongMaterial,
  ParallaxBarrierEffect,
  ParallaxShader,
  ParametricGeometries,
  ParametricGeometry,
  Pass,
  PeppersGhostEffect,
  PhysicalLightingModel,
  PixelShader,
  default39 as PointUVNode,
  PointerLockControls,
  default5 as PointsNodeMaterial,
  default40 as PositionNode,
  PositionalAudioHelper,
  ProgressiveLightMap,
  Projector,
  default25 as PropertyNode,
  RE_Direct_Physical,
  RGBELoader,
  RGBMLoader,
  RGBShiftShader,
  RectAreaLightHelper,
  RectAreaLightUniformsLib,
  default41 as ReferenceNode,
  default42 as ReflectNode,
  Reflector,
  ReflectorForSSRPass,
  ReflectorRTT,
  Refractor,
  RenderPass,
  RenderPixelatedPass,
  RenderableFace,
  RenderableLine,
  RenderableObject,
  RenderableSprite,
  RenderableVertex,
  Rhino3dmLoader,
  RollerCoasterGeometry,
  RollerCoasterLiftersGeometry,
  RollerCoasterShadowGeometry,
  RoomEnvironment,
  RoughnessMipmapper,
  RoundedBoxGeometry,
  SAOPass,
  SAOShader,
  SMAABlendShader,
  SMAAEdgesShader,
  SMAAPass,
  SMAAWeightsShader,
  SSAARenderPass,
  SSAOBlurShader,
  SSAODepthShader,
  SSAOPass,
  SSAOShader,
  SSRBlurShader,
  SSRDepthShader,
  SSRPass,
  SSRShader,
  STATE,
  STLExporter,
  STLLoader,
  SVGLoader,
  SVGObject,
  SVGRenderer,
  SavePass,
  SceneUtils,
  SelectionBox,
  SelectionHelper,
  SepiaShader,
  ShaderNode,
  ShaderPass,
  ShadowMapViewer,
  ShadowMesh,
  SimplexNoise,
  SimplifyModifier,
  SkeletonUtils,
  default43 as SkinningNode,
  Sky,
  SkyGeometry,
  SobelOperatorShader,
  default57 as SplitNode,
  default58 as SpriteSheetUVNode,
  StereoEffect,
  SubsurfaceScatteringShader,
  TAARenderPass,
  TDSLoader,
  TGALoader,
  TTFLoader,
  TeapotGeometry,
  TechnicolorShader,
  default26 as TempNode,
  TessellateModifier,
  TextGeometry as TextBufferGeometry,
  TextGeometry2 as TextGeometry,
  default44 as TextureNode,
  TexturePass,
  ThreeMFLoader,
  TiltLoader,
  default61 as TimerNode,
  ToneMapShader,
  ToonShader1,
  ToonShader2,
  ToonShaderDotted,
  ToonShaderHatching,
  TorusKnot,
  TrackballControls,
  TrackballControlsExp,
  TransformControls,
  TransformControlsGizmo,
  TransformControlsPlane,
  TreesGeometry,
  TrefoilKnot,
  TrefoilPolynomialKnot,
  TriangleBlurShader,
  TubePainter,
  USDZExporter,
  default45 as UVNode,
  UVsDebug,
  default27 as UniformNode,
  UnpackDepthRGBAShader,
  UnrealBloomPass,
  VOXDataTexture3D,
  VOXLoader,
  VOXMesh,
  VRButton,
  VRMLLoader,
  VRMLoader,
  VTKLoader,
  V_GGX_SmithCorrelated,
  default28 as VarNode,
  default29 as VaryNode,
  Vector2NodeUniform,
  Vector2Uniform,
  Vector3NodeUniform,
  Vector3Uniform,
  Vector4NodeUniform,
  Vector4Uniform,
  VertexNormalsHelper,
  VertexTangentsHelper,
  VerticalBlurShader,
  VerticalTiltShiftShader,
  VignetteShader,
  VivianiCurve,
  Volume,
  VolumeRenderShader1,
  VolumeSlice,
  Water,
  Water2,
  WaterRefractionShader,
  WebGPUSampled3DTexture,
  WebGPUSampledArrayTexture,
  WebGPUSampledCubeTexture,
  WebGPUSampledTexture,
  Wireframe,
  WireframeGeometry2,
  XLoader,
  XRControllerModelFactory,
  XREstimatedLight,
  XRHandMeshModel,
  XRHandModelFactory,
  XRHandPrimitiveModel,
  XYZLoader,
  abs,
  acos,
  add,
  addTo,
  alphaTest,
  and,
  asin,
  assign,
  atan,
  attribute,
  bitAnd,
  bitOr,
  bitXor,
  bmat3,
  bmat4,
  bool,
  buffer,
  bvec2,
  bvec3,
  bvec4,
  calcBSplineDerivatives,
  calcBSplinePoint,
  calcBasisFunctionDerivatives,
  calcBasisFunctions,
  calcKoverI,
  calcNURBSDerivatives,
  calcRationalCurveDerivatives,
  calcSurfacePoint,
  cameraPosition,
  ceil,
  clamp,
  color,
  computeMorphedAttributes,
  cond,
  cos,
  createText,
  cross,
  dFdx,
  dFdy,
  degrees,
  diffuseColor,
  distance,
  div,
  dot,
  edgeTable,
  element,
  equal,
  estimateBytesUsed,
  exp,
  exp2,
  faceforward,
  fetchProfile,
  fetchProfilesList,
  findSpan,
  float,
  floor,
  fract,
  fromType,
  getDistanceAttenuation,
  getErrorMessage,
  getUniforms,
  getWebGL2ErrorMessage,
  getWebGLErrorMessage,
  greaterThan,
  greaterThanEqual,
  imat3,
  imat4,
  initSplineTexture,
  int,
  interleaveAttributes,
  inversesqrt,
  invert,
  isWebGL2Available,
  isWebGLAvailable,
  ivec2,
  ivec3,
  ivec4,
  join,
  label,
  length,
  lessThan,
  lessThanEqual,
  log,
  log2,
  mat3,
  mat4,
  max,
  mergeBufferAttributes,
  mergeBufferGeometries,
  mergeVertices,
  metalness,
  min,
  mix,
  mod,
  modifyShader,
  mul,
  negate,
  nodeObject,
  normalGeometry,
  normalLocal,
  normalView,
  normalWorld,
  normalize,
  or,
  positionLocal,
  positionView,
  positionViewDirection,
  positionWorld,
  pow,
  pow2,
  pow3,
  pow4,
  radians,
  reflect,
  refract,
  remainder,
  roughness,
  round,
  sampler,
  saturate,
  scaleTo,
  shiftLeft,
  shiftRight,
  sign,
  sin,
  smoothstep,
  specularColor,
  sqrt,
  step,
  sub,
  tan,
  temp,
  texture,
  toTrianglesDrawMode,
  transformDirection,
  transformedNormalView,
  triTable,
  uint,
  umat3,
  umat4,
  uniform,
  updateSplineTexture,
  uv,
  uvec2,
  uvec3,
  uvec4,
  vec2,
  vec3,
  vec4,
  viewMatrix,
  xor
};
