// three-stdlib/index.js
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
require("./offscreen/scene.js");
require("./offscreen/offscreen.js");
require("./offscreen/jank.js");
const MD2CharacterComplex = require("./misc/MD2CharacterComplex.js");
const ConvexObjectBreaker = require("./misc/ConvexObjectBreaker.js");
const MorphBlendMesh = require("./misc/MorphBlendMesh.js");
const GPUComputationRenderer = require("./misc/GPUComputationRenderer.js");
const Gyroscope = require("./misc/Gyroscope.js");
const MorphAnimMesh = require("./misc/MorphAnimMesh.js");
const RollerCoaster = require("./misc/RollerCoaster.js");
const WebGL = require("./misc/WebGL.js");
const MD2Character = require("./misc/MD2Character.js");
const VolumeSlice = require("./misc/VolumeSlice.js");
const TubePainter = require("./misc/TubePainter.js");
const Volume = require("./misc/Volume.js");
const ProgressiveLightmap = require("./misc/ProgressiveLightmap.js");
const CSS2DRenderer = require("./renderers/CSS2DRenderer.js");
require("./renderers/nodes/core/constants.js");
require("./renderers/nodes/core/NodeBuilder.js");
require("./renderers/nodes/core/InputNode.js");
require("./renderers/nodes/core/NodeUniform.js");
require("./renderers/nodes/core/NodeVary.js");
require("./renderers/nodes/core/NodeFrame.js");
require("./renderers/nodes/core/Node.js");
require("./renderers/nodes/core/NodeSlot.js");
require("./renderers/nodes/core/NodeAttribute.js");
require("./renderers/nodes/core/VaryNode.js");
require("./renderers/nodes/core/AttributeNode.js");
require("./renderers/nodes/accessors/CameraNode.js");
require("./renderers/nodes/accessors/UVNode.js");
require("./renderers/nodes/accessors/ModelNode.js");
require("./renderers/nodes/accessors/NormalNode.js");
require("./renderers/nodes/accessors/PositionNode.js");
require("./renderers/nodes/accessors/ModelViewProjectionNode.js");
require("./renderers/nodes/utils/SwitchNode.js");
require("./renderers/nodes/utils/TimerNode.js");
require("./renderers/nodes/math/OperatorNode.js");
require("./renderers/nodes/math/MathNode.js");
require("./renderers/nodes/inputs/ColorNode.js");
require("./renderers/nodes/inputs/Vector3Node.js");
require("./renderers/nodes/inputs/Matrix4Node.js");
require("./renderers/nodes/inputs/TextureNode.js");
require("./renderers/nodes/inputs/FloatNode.js");
require("./renderers/nodes/inputs/Vector2Node.js");
require("./renderers/nodes/inputs/Vector4Node.js");
require("./renderers/nodes/inputs/Matrix3Node.js");
const constants = require("./renderers/webgpu/constants.js");
require("./renderers/webgpu/WebGPUTextures.js");
require("./renderers/webgpu/WebGPUInfo.js");
require("./renderers/webgpu/WebGPURenderLists.js");
require("./renderers/webgpu/WebGPUTextureRenderer.js");
require("./renderers/webgpu/WebGPURenderPipelines.js");
require("./renderers/webgpu/WebGPUGeometries.js");
require("./renderers/webgpu/nodes/WebGPUNodeBuilder.js");
const WebGPUNodeUniform = require("./renderers/webgpu/nodes/WebGPUNodeUniform.js");
require("./renderers/webgpu/nodes/WebGPUNodeUniformsGroup.js");
require("./renderers/webgpu/nodes/ShaderLib.js");
require("./renderers/webgpu/nodes/WebGPUNodes.js");
require("./renderers/webgpu/WebGPUBindings.js");
const WebGPUSampledTexture = require("./renderers/webgpu/WebGPUSampledTexture.js");
require("./renderers/webgpu/WebGPU.js");
require("./renderers/webgpu/WebGPUUniformsGroup.js");
require("./renderers/webgpu/WebGPURenderer.js");
require("./renderers/webgpu/WebGPUComputePipelines.js");
const WebGPUUniform = require("./renderers/webgpu/WebGPUUniform.js");
require("./renderers/webgpu/WebGPUObjects.js");
require("./renderers/webgpu/WebGPUSampler.js");
require("./renderers/webgpu/WebGPUBackground.js");
require("./renderers/webgpu/WebGPUBinding.js");
require("./renderers/webgpu/WebGPUProperties.js");
require("./renderers/webgpu/WebGPUAttributes.js");
require("./renderers/webgpu/WebGPUStorageBuffer.js");
require("./renderers/webgpu/WebGPUTextureUtils.js");
const CSS3DRenderer = require("./renderers/CSS3DRenderer.js");
const Projector = require("./renderers/Projector.js");
const SVGRenderer = require("./renderers/SVGRenderer.js");
const FlakesTexture = require("./textures/FlakesTexture.js");
const CurveModifier = require("./modifiers/CurveModifier.js");
const SimplifyModifier = require("./modifiers/SimplifyModifier.js");
const EdgeSplitModifier = require("./modifiers/EdgeSplitModifier.js");
const TessellateModifier = require("./modifiers/TessellateModifier.js");
const Nodes = require("./nodes/Nodes.js");
const GLTFExporter = require("./exporters/GLTFExporter.js");
const USDZExporter = require("./exporters/USDZExporter.js");
const PLYExporter = require("./exporters/PLYExporter.js");
const DRACOExporter = require("./exporters/DRACOExporter.js");
const ColladaExporter = require("./exporters/ColladaExporter.js");
const MMDExporter = require("./exporters/MMDExporter.js");
const STLExporter = require("./exporters/STLExporter.js");
const OBJExporter = require("./exporters/OBJExporter.js");
const RoomEnvironment = require("./environments/RoomEnvironment.js");
const AnimationClipCreator = require("./animation/AnimationClipCreator.js");
const CCDIKSolver = require("./animation/CCDIKSolver.js");
const MMDPhysics = require("./animation/MMDPhysics.js");
const MMDAnimationHelper = require("./animation/MMDAnimationHelper.js");
const Reflector = require("./objects/Reflector.js");
const Refractor = require("./objects/Refractor.js");
const ShadowMesh = require("./objects/ShadowMesh.js");
const Lensflare = require("./objects/Lensflare.js");
const Water = require("./objects/Water.js");
const MarchingCubes = require("./objects/MarchingCubes.js");
const LightningStorm = require("./objects/LightningStorm.js");
const ReflectorRTT = require("./objects/ReflectorRTT.js");
const ReflectorForSSRPass = require("./objects/ReflectorForSSRPass.js");
const Sky = require("./objects/Sky.js");
const Water2 = require("./objects/Water2.js");
const GroundProjectedEnv = require("./objects/GroundProjectedEnv.js");
const SceneUtils = require("./utils/SceneUtils.js");
const UVsDebug = require("./utils/UVsDebug.js");
const GeometryUtils = require("./utils/GeometryUtils.js");
const RoughnessMipmapper = require("./utils/RoughnessMipmapper.js");
const SkeletonUtils = require("./utils/SkeletonUtils.js");
const ShadowMapViewer = require("./utils/ShadowMapViewer.js");
const BufferGeometryUtils = require("./utils/BufferGeometryUtils.js");
const GeometryCompressionUtils = require("./utils/GeometryCompressionUtils.js");
const CinematicCamera = require("./cameras/CinematicCamera.js");
const ConvexHull = require("./math/ConvexHull.js");
const MeshSurfaceSampler = require("./math/MeshSurfaceSampler.js");
const SimplexNoise = require("./math/SimplexNoise.js");
const OBB = require("./math/OBB.js");
const Capsule = require("./math/Capsule.js");
const ColorConverter = require("./math/ColorConverter.js");
const ImprovedNoise = require("./math/ImprovedNoise.js");
const Octree = require("./math/Octree.js");
const Lut = require("./math/Lut.js");
const CameraControls = require("./controls/experimental/CameraControls.js");
const FirstPersonControls = require("./controls/FirstPersonControls.js");
const TransformControls = require("./controls/TransformControls.js");
const DragControls = require("./controls/DragControls.js");
const PointerLockControls = require("./controls/PointerLockControls.js");
const DeviceOrientationControls = require("./controls/DeviceOrientationControls.js");
const TrackballControls = require("./controls/TrackballControls.js");
const OrbitControls = require("./controls/OrbitControls.js");
const ArcballControls = require("./controls/ArcballControls.js");
const FlyControls = require("./controls/FlyControls.js");
const LUTPass = require("./postprocessing/LUTPass.js");
const ClearPass = require("./postprocessing/ClearPass.js");
const GlitchPass = require("./postprocessing/GlitchPass.js");
const HalftonePass = require("./postprocessing/HalftonePass.js");
const SMAAPass = require("./postprocessing/SMAAPass.js");
const FilmPass = require("./postprocessing/FilmPass.js");
const OutlinePass = require("./postprocessing/OutlinePass.js");
const SSAOPass = require("./postprocessing/SSAOPass.js");
const SavePass = require("./postprocessing/SavePass.js");
const BokehPass = require("./postprocessing/BokehPass.js");
const Pass = require("./postprocessing/Pass.js");
const TexturePass = require("./postprocessing/TexturePass.js");
const AdaptiveToneMappingPass = require("./postprocessing/AdaptiveToneMappingPass.js");
const UnrealBloomPass = require("./postprocessing/UnrealBloomPass.js");
const CubeTexturePass = require("./postprocessing/CubeTexturePass.js");
const SAOPass = require("./postprocessing/SAOPass.js");
const AfterimagePass = require("./postprocessing/AfterimagePass.js");
const MaskPass = require("./postprocessing/MaskPass.js");
const EffectComposer = require("./postprocessing/EffectComposer.js");
const DotScreenPass = require("./postprocessing/DotScreenPass.js");
const SSRPass = require("./postprocessing/SSRPass.js");
const TAARenderPass = require("./postprocessing/TAARenderPass.js");
const ShaderPass = require("./postprocessing/ShaderPass.js");
const SSAARenderPass = require("./postprocessing/SSAARenderPass.js");
const RenderPass = require("./postprocessing/RenderPass.js");
const RenderPixelatedPass = require("./postprocessing/RenderPixelatedPass.js");
const BloomPass = require("./postprocessing/BloomPass.js");
const ARButton = require("./webxr/ARButton.js");
const OculusHandModel = require("./webxr/OculusHandModel.js");
const OculusHandPointerModel = require("./webxr/OculusHandPointerModel.js");
const Text2D = require("./webxr/Text2D.js");
const VRButton = require("./webxr/VRButton.js");
const XRControllerModelFactory = require("./webxr/XRControllerModelFactory.js");
const XREstimatedLight = require("./webxr/XREstimatedLight.js");
const XRHandMeshModel = require("./webxr/XRHandMeshModel.js");
const XRHandModelFactory = require("./webxr/XRHandModelFactory.js");
const XRHandPrimitiveModel = require("./webxr/XRHandPrimitiveModel.js");
const ParametricGeometries = require("./geometries/ParametricGeometries.js");
const ParametricGeometry = require("./geometries/ParametricGeometry.js");
const ConvexGeometry = require("./geometries/ConvexGeometry.js");
const LightningStrike = require("./geometries/LightningStrike.js");
const RoundedBoxGeometry = require("./geometries/RoundedBoxGeometry.js");
const BoxLineGeometry = require("./geometries/BoxLineGeometry.js");
const DecalGeometry = require("./geometries/DecalGeometry.js");
const TeapotGeometry = require("./geometries/TeapotGeometry.js");
const TextGeometry = require("./geometries/TextGeometry.js");
const CSM = require("./csm/CSM.js");
const CSMFrustum = require("./csm/CSMFrustum.js");
const CSMHelper = require("./csm/CSMHelper.js");
const CSMShader = require("./csm/CSMShader.js");
require("./shaders/types.js");
const ACESFilmicToneMappingShader = require("./shaders/ACESFilmicToneMappingShader.js");
const AfterimageShader = require("./shaders/AfterimageShader.js");
const BasicShader = require("./shaders/BasicShader.js");
const BleachBypassShader = require("./shaders/BleachBypassShader.js");
const BlendShader = require("./shaders/BlendShader.js");
const BokehShader = require("./shaders/BokehShader.js");
const BokehShader2 = require("./shaders/BokehShader2.js");
const BrightnessContrastShader = require("./shaders/BrightnessContrastShader.js");
const ColorCorrectionShader = require("./shaders/ColorCorrectionShader.js");
const ColorifyShader = require("./shaders/ColorifyShader.js");
const ConvolutionShader = require("./shaders/ConvolutionShader.js");
const CopyShader = require("./shaders/CopyShader.js");
const DOFMipMapShader = require("./shaders/DOFMipMapShader.js");
const DepthLimitedBlurShader = require("./shaders/DepthLimitedBlurShader.js");
const DigitalGlitch = require("./shaders/DigitalGlitch.js");
const DotScreenShader = require("./shaders/DotScreenShader.js");
const FXAAShader = require("./shaders/FXAAShader.js");
const FilmShader = require("./shaders/FilmShader.js");
const FocusShader = require("./shaders/FocusShader.js");
const FreiChenShader = require("./shaders/FreiChenShader.js");
const FresnelShader = require("./shaders/FresnelShader.js");
const GammaCorrectionShader = require("./shaders/GammaCorrectionShader.js");
const GodRaysShader = require("./shaders/GodRaysShader.js");
const HalftoneShader = require("./shaders/HalftoneShader.js");
const HorizontalBlurShader = require("./shaders/HorizontalBlurShader.js");
const HorizontalTiltShiftShader = require("./shaders/HorizontalTiltShiftShader.js");
const HueSaturationShader = require("./shaders/HueSaturationShader.js");
const KaleidoShader = require("./shaders/KaleidoShader.js");
const LuminosityHighPassShader = require("./shaders/LuminosityHighPassShader.js");
const LuminosityShader = require("./shaders/LuminosityShader.js");
const MirrorShader = require("./shaders/MirrorShader.js");
const NormalMapShader = require("./shaders/NormalMapShader.js");
const ParallaxShader = require("./shaders/ParallaxShader.js");
const PixelShader = require("./shaders/PixelShader.js");
const RGBShiftShader = require("./shaders/RGBShiftShader.js");
const SAOShader = require("./shaders/SAOShader.js");
const SMAAShader = require("./shaders/SMAAShader.js");
const SSAOShader = require("./shaders/SSAOShader.js");
const SSRShader = require("./shaders/SSRShader.js");
const SepiaShader = require("./shaders/SepiaShader.js");
const SobelOperatorShader = require("./shaders/SobelOperatorShader.js");
const SubsurfaceScatteringShader = require("./shaders/SubsurfaceScatteringShader.js");
const TechnicolorShader = require("./shaders/TechnicolorShader.js");
const ToneMapShader = require("./shaders/ToneMapShader.js");
const ToonShader = require("./shaders/ToonShader.js");
const TriangleBlurShader = require("./shaders/TriangleBlurShader.js");
const UnpackDepthRGBAShader = require("./shaders/UnpackDepthRGBAShader.js");
const VerticalBlurShader = require("./shaders/VerticalBlurShader.js");
const VerticalTiltShiftShader = require("./shaders/VerticalTiltShiftShader.js");
const VignetteShader = require("./shaders/VignetteShader.js");
const VolumeShader = require("./shaders/VolumeShader.js");
const WaterRefractionShader = require("./shaders/WaterRefractionShader.js");
const SelectionHelper = require("./interactive/SelectionHelper.js");
const SelectionBox = require("./interactive/SelectionBox.js");
const AmmoPhysics = require("./physics/AmmoPhysics.js");
const ParallaxBarrierEffect = require("./effects/ParallaxBarrierEffect.js");
const PeppersGhostEffect = require("./effects/PeppersGhostEffect.js");
const OutlineEffect = require("./effects/OutlineEffect.js");
const AnaglyphEffect = require("./effects/AnaglyphEffect.js");
const AsciiEffect = require("./effects/AsciiEffect.js");
const StereoEffect = require("./effects/StereoEffect.js");
const FBXLoader = require("./loaders/FBXLoader.js");
const FontLoader = require("./loaders/FontLoader.js");
const TGALoader = require("./loaders/TGALoader.js");
const LUTCubeLoader = require("./loaders/LUTCubeLoader.js");
const NRRDLoader = require("./loaders/NRRDLoader.js");
const STLLoader = require("./loaders/STLLoader.js");
const MTLLoader = require("./loaders/MTLLoader.js");
const XLoader = require("./loaders/XLoader.js");
const BVHLoader = require("./loaders/BVHLoader.js");
const KMZLoader = require("./loaders/KMZLoader.js");
const VRMLoader = require("./loaders/VRMLoader.js");
const VRMLLoader = require("./loaders/VRMLLoader.js");
const KTX2Loader = require("./loaders/KTX2Loader.js");
const LottieLoader = require("./loaders/LottieLoader.js");
const TTFLoader = require("./loaders/TTFLoader.js");
const RGBELoader = require("./loaders/RGBELoader.js");
const AssimpLoader = require("./loaders/AssimpLoader.js");
const ColladaLoader = require("./loaders/ColladaLoader.js");
const MDDLoader = require("./loaders/MDDLoader.js");
const EXRLoader = require("./loaders/EXRLoader.js");
const _3MFLoader = require("./loaders/3MFLoader.js");
const XYZLoader = require("./loaders/XYZLoader.js");
const VTKLoader = require("./loaders/VTKLoader.js");
const LUT3dlLoader = require("./loaders/LUT3dlLoader.js");
const DDSLoader = require("./loaders/DDSLoader.js");
const PVRLoader = require("./loaders/PVRLoader.js");
const GCodeLoader = require("./loaders/GCodeLoader.js");
const BasisTextureLoader = require("./loaders/BasisTextureLoader.js");
const TDSLoader = require("./loaders/TDSLoader.js");
const LDrawLoader = require("./loaders/LDrawLoader.js");
const GLTFLoader = require("./loaders/GLTFLoader.js");
const LWO3Parser = require("./loaders/lwo/LWO3Parser.js");
const LWO2Parser = require("./loaders/lwo/LWO2Parser.js");
const IFFParser = require("./loaders/lwo/IFFParser.js");
const SVGLoader = require("./loaders/SVGLoader.js");
const _3DMLoader = require("./loaders/3DMLoader.js");
const OBJLoader = require("./loaders/OBJLoader.js");
const AMFLoader = require("./loaders/AMFLoader.js");
const MMDLoader = require("./loaders/MMDLoader.js");
const MD2Loader = require("./loaders/MD2Loader.js");
const KTXLoader = require("./loaders/KTXLoader.js");
const TiltLoader = require("./loaders/TiltLoader.js");
const DRACOLoader = require("./loaders/DRACOLoader.js");
const HDRCubeTextureLoader = require("./loaders/HDRCubeTextureLoader.js");
const PDBLoader = require("./loaders/PDBLoader.js");
const PRWMLoader = require("./loaders/PRWMLoader.js");
const RGBMLoader = require("./loaders/RGBMLoader.js");
const VOXLoader = require("./loaders/VOXLoader.js");
const NodeMaterialLoader = require("./loaders/NodeMaterialLoader.js");
const PCDLoader = require("./loaders/PCDLoader.js");
const LWOLoader = require("./loaders/LWOLoader.js");
const PLYLoader = require("./loaders/PLYLoader.js");
const LineSegmentsGeometry = require("./lines/LineSegmentsGeometry.js");
const LineGeometry = require("./lines/LineGeometry.js");
const Wireframe = require("./lines/Wireframe.js");
const WireframeGeometry2 = require("./lines/WireframeGeometry2.js");
const Line2 = require("./lines/Line2.js");
const LineMaterial = require("./lines/LineMaterial.js");
const LineSegments2 = require("./lines/LineSegments2.js");
const LightProbeHelper = require("./helpers/LightProbeHelper.js");
const VertexTangentsHelper = require("./helpers/VertexTangentsHelper.js");
const PositionalAudioHelper = require("./helpers/PositionalAudioHelper.js");
const VertexNormalsHelper = require("./helpers/VertexNormalsHelper.js");
const RectAreaLightHelper = require("./helpers/RectAreaLightHelper.js");
const RectAreaLightUniformsLib = require("./lights/RectAreaLightUniformsLib.js");
const LightProbeGenerator = require("./lights/LightProbeGenerator.js");
const NURBSUtils = require("./curves/NURBSUtils.js");
const NURBSCurve = require("./curves/NURBSCurve.js");
const NURBSSurface = require("./curves/NURBSSurface.js");
const CurveExtras = require("./curves/CurveExtras.js");
const Geometry = require("./deprecated/Geometry.js");
const MeshoptDecoder = require("./libs/MeshoptDecoder.js");
const MotionControllers = require("./libs/MotionControllers.js");
const BSDFs = require("./nodes/functions/BSDFs.js");
const LineBasicNodeMaterial = require("./nodes/materials/LineBasicNodeMaterial.js");
const MeshBasicNodeMaterial = require("./nodes/materials/MeshBasicNodeMaterial.js");
const MeshStandardNodeMaterial = require("./nodes/materials/MeshStandardNodeMaterial.js");
const PointsNodeMaterial = require("./nodes/materials/PointsNodeMaterial.js");
const ShaderNode = require("./nodes/ShaderNode.js");
const ArrayUniformNode = require("./nodes/core/ArrayUniformNode.js");
const AttributeNode = require("./nodes/core/AttributeNode.js");
const BypassNode = require("./nodes/core/BypassNode.js");
const CodeNode = require("./nodes/core/CodeNode.js");
const ContextNode = require("./nodes/core/ContextNode.js");
const ConstNode = require("./nodes/core/ConstNode.js");
const ExpressionNode = require("./nodes/core/ExpressionNode.js");
const FunctionCallNode = require("./nodes/core/FunctionCallNode.js");
const FunctionNode = require("./nodes/core/FunctionNode.js");
const Node = require("./nodes/core/Node.js");
const NodeAttribute = require("./nodes/core/NodeAttribute.js");
const NodeBuilder = require("./nodes/core/NodeBuilder.js");
const NodeCode = require("./nodes/core/NodeCode.js");
const NodeFrame = require("./nodes/core/NodeFrame.js");
const NodeFunctionInput = require("./nodes/core/NodeFunctionInput.js");
const NodeKeywords = require("./nodes/core/NodeKeywords.js");
const NodeUniform = require("./nodes/core/NodeUniform.js");
const NodeVar = require("./nodes/core/NodeVar.js");
const NodeVary = require("./nodes/core/NodeVary.js");
const PropertyNode = require("./nodes/core/PropertyNode.js");
const TempNode = require("./nodes/core/TempNode.js");
const UniformNode = require("./nodes/core/UniformNode.js");
const VarNode = require("./nodes/core/VarNode.js");
const VaryNode = require("./nodes/core/VaryNode.js");
const BufferNode = require("./nodes/accessors/BufferNode.js");
const CameraNode = require("./nodes/accessors/CameraNode.js");
const CubeTextureNode = require("./nodes/accessors/CubeTextureNode.js");
const MaterialNode = require("./nodes/accessors/MaterialNode.js");
const MaterialReferenceNode = require("./nodes/accessors/MaterialReferenceNode.js");
const ModelNode = require("./nodes/accessors/ModelNode.js");
const ModelViewProjectionNode = require("./nodes/accessors/ModelViewProjectionNode.js");
const NormalNode = require("./nodes/accessors/NormalNode.js");
const Object3DNode = require("./nodes/accessors/Object3DNode.js");
const PointUVNode = require("./nodes/accessors/PointUVNode.js");
const PositionNode = require("./nodes/accessors/PositionNode.js");
const ReferenceNode = require("./nodes/accessors/ReferenceNode.js");
const ReflectNode = require("./nodes/accessors/ReflectNode.js");
const SkinningNode = require("./nodes/accessors/SkinningNode.js");
const TextureNode = require("./nodes/accessors/TextureNode.js");
const UVNode = require("./nodes/accessors/UVNode.js");
const ColorSpaceNode = require("./nodes/display/ColorSpaceNode.js");
const NormalMapNode = require("./nodes/display/NormalMapNode.js");
const MathNode = require("./nodes/math/MathNode.js");
const OperatorNode = require("./nodes/math/OperatorNode.js");
const CondNode = require("./nodes/math/CondNode.js");
const LightContextNode = require("./nodes/lights/LightContextNode.js");
const LightNode = require("./nodes/lights/LightNode.js");
const LightsNode = require("./nodes/lights/LightsNode.js");
const ArrayElementNode = require("./nodes/utils/ArrayElementNode.js");
const ConvertNode = require("./nodes/utils/ConvertNode.js");
const JoinNode = require("./nodes/utils/JoinNode.js");
const SplitNode = require("./nodes/utils/SplitNode.js");
const SpriteSheetUVNode = require("./nodes/utils/SpriteSheetUVNode.js");
const MatcapUVNode = require("./nodes/utils/MatcapUVNode.js");
const OscNode = require("./nodes/utils/OscNode.js");
const TimerNode = require("./nodes/utils/TimerNode.js");
const CheckerNode = require("./nodes/procedural/CheckerNode.js");
const FogNode = require("./nodes/fog/FogNode.js");
const FogRangeNode = require("./nodes/fog/FogRangeNode.js");
const NodeLoader = require("./nodes/loaders/NodeLoader.js");
const NodeObjectLoader = require("./nodes/loaders/NodeObjectLoader.js");
exports.MD2CharacterComplex = MD2CharacterComplex.MD2CharacterComplex;
exports.ConvexObjectBreaker = ConvexObjectBreaker.ConvexObjectBreaker;
exports.MorphBlendMesh = MorphBlendMesh.MorphBlendMesh;
exports.GPUComputationRenderer = GPUComputationRenderer.GPUComputationRenderer;
exports.Gyroscope = Gyroscope.Gyroscope;
exports.MorphAnimMesh = MorphAnimMesh.MorphAnimMesh;
exports.RollerCoasterGeometry = RollerCoaster.RollerCoasterGeometry;
exports.RollerCoasterLiftersGeometry = RollerCoaster.RollerCoasterLiftersGeometry;
exports.RollerCoasterShadowGeometry = RollerCoaster.RollerCoasterShadowGeometry;
exports.SkyGeometry = RollerCoaster.SkyGeometry;
exports.TreesGeometry = RollerCoaster.TreesGeometry;
exports.getErrorMessage = WebGL.getErrorMessage;
exports.getWebGL2ErrorMessage = WebGL.getWebGL2ErrorMessage;
exports.getWebGLErrorMessage = WebGL.getWebGLErrorMessage;
exports.isWebGL2Available = WebGL.isWebGL2Available;
exports.isWebGLAvailable = WebGL.isWebGLAvailable;
exports.MD2Character = MD2Character.MD2Character;
exports.VolumeSlice = VolumeSlice.VolumeSlice;
exports.TubePainter = TubePainter.TubePainter;
exports.Volume = Volume.Volume;
exports.ProgressiveLightMap = ProgressiveLightmap.ProgressiveLightMap;
exports.CSS2DObject = CSS2DRenderer.CSS2DObject;
exports.CSS2DRenderer = CSS2DRenderer.CSS2DRenderer;
exports.BlendColorFactor = constants.BlendColorFactor;
exports.GPUAddressMode = constants.GPUAddressMode;
exports.GPUBindingType = constants.GPUBindingType;
exports.GPUBlendFactor = constants.GPUBlendFactor;
exports.GPUBlendOperation = constants.GPUBlendOperation;
exports.GPUColorWriteFlags = constants.GPUColorWriteFlags;
exports.GPUCompareFunction = constants.GPUCompareFunction;
exports.GPUCullMode = constants.GPUCullMode;
exports.GPUFilterMode = constants.GPUFilterMode;
exports.GPUFrontFace = constants.GPUFrontFace;
exports.GPUIndexFormat = constants.GPUIndexFormat;
exports.GPUInputStepMode = constants.GPUInputStepMode;
exports.GPULoadOp = constants.GPULoadOp;
exports.GPUPrimitiveTopology = constants.GPUPrimitiveTopology;
exports.GPUStencilOperation = constants.GPUStencilOperation;
exports.GPUStoreOp = constants.GPUStoreOp;
exports.GPUTextureDimension = constants.GPUTextureDimension;
exports.GPUTextureFormat = constants.GPUTextureFormat;
exports.GPUTextureViewDimension = constants.GPUTextureViewDimension;
exports.GPUVertexFormat = constants.GPUVertexFormat;
exports.OneMinusBlendColorFactor = constants.OneMinusBlendColorFactor;
exports.ColorNodeUniform = WebGPUNodeUniform.ColorNodeUniform;
exports.FloatNodeUniform = WebGPUNodeUniform.FloatNodeUniform;
exports.Matrix3NodeUniform = WebGPUNodeUniform.Matrix3NodeUniform;
exports.Matrix4NodeUniform = WebGPUNodeUniform.Matrix4NodeUniform;
exports.Vector2NodeUniform = WebGPUNodeUniform.Vector2NodeUniform;
exports.Vector3NodeUniform = WebGPUNodeUniform.Vector3NodeUniform;
exports.Vector4NodeUniform = WebGPUNodeUniform.Vector4NodeUniform;
exports.WebGPUSampled3DTexture = WebGPUSampledTexture.WebGPUSampled3DTexture;
exports.WebGPUSampledArrayTexture = WebGPUSampledTexture.WebGPUSampledArrayTexture;
exports.WebGPUSampledCubeTexture = WebGPUSampledTexture.WebGPUSampledCubeTexture;
exports.WebGPUSampledTexture = WebGPUSampledTexture.WebGPUSampledTexture;
exports.ColorUniform = WebGPUUniform.ColorUniform;
exports.FloatUniform = WebGPUUniform.FloatUniform;
exports.Matrix3Uniform = WebGPUUniform.Matrix3Uniform;
exports.Matrix4Uniform = WebGPUUniform.Matrix4Uniform;
exports.Vector2Uniform = WebGPUUniform.Vector2Uniform;
exports.Vector3Uniform = WebGPUUniform.Vector3Uniform;
exports.Vector4Uniform = WebGPUUniform.Vector4Uniform;
exports.CSS3DObject = CSS3DRenderer.CSS3DObject;
exports.CSS3DRenderer = CSS3DRenderer.CSS3DRenderer;
exports.CSS3DSprite = CSS3DRenderer.CSS3DSprite;
exports.Projector = Projector.Projector;
exports.RenderableFace = Projector.RenderableFace;
exports.RenderableLine = Projector.RenderableLine;
exports.RenderableObject = Projector.RenderableObject;
exports.RenderableSprite = Projector.RenderableSprite;
exports.RenderableVertex = Projector.RenderableVertex;
exports.SVGObject = SVGRenderer.SVGObject;
exports.SVGRenderer = SVGRenderer.SVGRenderer;
exports.FlakesTexture = FlakesTexture.FlakesTexture;
exports.Flow = CurveModifier.Flow;
exports.InstancedFlow = CurveModifier.InstancedFlow;
exports.getUniforms = CurveModifier.getUniforms;
exports.initSplineTexture = CurveModifier.initSplineTexture;
exports.modifyShader = CurveModifier.modifyShader;
exports.updateSplineTexture = CurveModifier.updateSplineTexture;
exports.SimplifyModifier = SimplifyModifier.SimplifyModifier;
exports.EdgeSplitModifier = EdgeSplitModifier.EdgeSplitModifier;
exports.TessellateModifier = TessellateModifier.TessellateModifier;
exports.fromType = Nodes.fromType;
exports.GLTFExporter = GLTFExporter.GLTFExporter;
exports.USDZExporter = USDZExporter.USDZExporter;
exports.PLYExporter = PLYExporter.PLYExporter;
exports.DRACOExporter = DRACOExporter.DRACOExporter;
exports.ColladaExporter = ColladaExporter.ColladaExporter;
exports.MMDExporter = MMDExporter.MMDExporter;
exports.STLExporter = STLExporter.STLExporter;
exports.OBJExporter = OBJExporter.OBJExporter;
exports.RoomEnvironment = RoomEnvironment.RoomEnvironment;
exports.AnimationClipCreator = AnimationClipCreator.AnimationClipCreator;
exports.CCDIKHelper = CCDIKSolver.CCDIKHelper;
exports.CCDIKSolver = CCDIKSolver.CCDIKSolver;
exports.MMDPhysics = MMDPhysics.MMDPhysics;
exports.MMDAnimationHelper = MMDAnimationHelper.MMDAnimationHelper;
exports.Reflector = Reflector.Reflector;
exports.Refractor = Refractor.Refractor;
exports.ShadowMesh = ShadowMesh.ShadowMesh;
exports.Lensflare = Lensflare.Lensflare;
exports.LensflareElement = Lensflare.LensflareElement;
exports.Water = Water.Water;
exports.MarchingCubes = MarchingCubes.MarchingCubes;
exports.edgeTable = MarchingCubes.edgeTable;
exports.triTable = MarchingCubes.triTable;
exports.LightningStorm = LightningStorm.LightningStorm;
exports.ReflectorRTT = ReflectorRTT.ReflectorRTT;
exports.ReflectorForSSRPass = ReflectorForSSRPass.ReflectorForSSRPass;
exports.Sky = Sky.Sky;
exports.Water2 = Water2.Water2;
exports.GroundProjectedEnv = GroundProjectedEnv.GroundProjectedEnv;
exports.SceneUtils = SceneUtils.SceneUtils;
exports.UVsDebug = UVsDebug.UVsDebug;
exports.GeometryUtils = GeometryUtils.GeometryUtils;
exports.RoughnessMipmapper = RoughnessMipmapper.RoughnessMipmapper;
exports.SkeletonUtils = SkeletonUtils.SkeletonUtils;
exports.ShadowMapViewer = ShadowMapViewer.ShadowMapViewer;
exports.computeMorphedAttributes = BufferGeometryUtils.computeMorphedAttributes;
exports.estimateBytesUsed = BufferGeometryUtils.estimateBytesUsed;
exports.interleaveAttributes = BufferGeometryUtils.interleaveAttributes;
exports.mergeBufferAttributes = BufferGeometryUtils.mergeBufferAttributes;
exports.mergeBufferGeometries = BufferGeometryUtils.mergeBufferGeometries;
exports.mergeVertices = BufferGeometryUtils.mergeVertices;
exports.toTrianglesDrawMode = BufferGeometryUtils.toTrianglesDrawMode;
exports.GeometryCompressionUtils = GeometryCompressionUtils.GeometryCompressionUtils;
exports.PackedPhongMaterial = GeometryCompressionUtils.PackedPhongMaterial;
exports.CinematicCamera = CinematicCamera.CinematicCamera;
exports.ConvexHull = ConvexHull.ConvexHull;
exports.MeshSurfaceSampler = MeshSurfaceSampler.MeshSurfaceSampler;
exports.SimplexNoise = SimplexNoise.SimplexNoise;
exports.OBB = OBB.OBB;
exports.Capsule = Capsule.Capsule;
exports.ColorConverter = ColorConverter.ColorConverter;
exports.ImprovedNoise = ImprovedNoise.ImprovedNoise;
exports.Octree = Octree.Octree;
exports.ColorMapKeywords = Lut.ColorMapKeywords;
exports.Lut = Lut.Lut;
exports.CameraControls = CameraControls.CameraControls;
exports.MapControlsExp = CameraControls.MapControlsExp;
exports.OrbitControlsExp = CameraControls.OrbitControlsExp;
exports.STATE = CameraControls.STATE;
exports.TrackballControlsExp = CameraControls.TrackballControlsExp;
exports.FirstPersonControls = FirstPersonControls.FirstPersonControls;
exports.TransformControls = TransformControls.TransformControls;
exports.TransformControlsGizmo = TransformControls.TransformControlsGizmo;
exports.TransformControlsPlane = TransformControls.TransformControlsPlane;
exports.DragControls = DragControls.DragControls;
exports.PointerLockControls = PointerLockControls.PointerLockControls;
exports.DeviceOrientationControls = DeviceOrientationControls.DeviceOrientationControls;
exports.TrackballControls = TrackballControls.TrackballControls;
exports.MapControls = OrbitControls.MapControls;
exports.OrbitControls = OrbitControls.OrbitControls;
exports.ArcballControls = ArcballControls.ArcballControls;
exports.FlyControls = FlyControls.FlyControls;
exports.LUTPass = LUTPass.LUTPass;
exports.ClearPass = ClearPass.ClearPass;
exports.GlitchPass = GlitchPass.GlitchPass;
exports.HalftonePass = HalftonePass.HalftonePass;
exports.SMAAPass = SMAAPass.SMAAPass;
exports.FilmPass = FilmPass.FilmPass;
exports.OutlinePass = OutlinePass.OutlinePass;
exports.SSAOPass = SSAOPass.SSAOPass;
exports.SavePass = SavePass.SavePass;
exports.BokehPass = BokehPass.BokehPass;
exports.FullScreenQuad = Pass.FullScreenQuad;
exports.Pass = Pass.Pass;
exports.TexturePass = TexturePass.TexturePass;
exports.AdaptiveToneMappingPass = AdaptiveToneMappingPass.AdaptiveToneMappingPass;
exports.UnrealBloomPass = UnrealBloomPass.UnrealBloomPass;
exports.CubeTexturePass = CubeTexturePass.CubeTexturePass;
exports.SAOPass = SAOPass.SAOPass;
exports.AfterimagePass = AfterimagePass.AfterimagePass;
exports.ClearMaskPass = MaskPass.ClearMaskPass;
exports.MaskPass = MaskPass.MaskPass;
exports.EffectComposer = EffectComposer.EffectComposer;
exports.DotScreenPass = DotScreenPass.DotScreenPass;
exports.SSRPass = SSRPass.SSRPass;
exports.TAARenderPass = TAARenderPass.TAARenderPass;
exports.ShaderPass = ShaderPass.ShaderPass;
exports.SSAARenderPass = SSAARenderPass.SSAARenderPass;
exports.RenderPass = RenderPass.RenderPass;
exports.RenderPixelatedPass = RenderPixelatedPass.RenderPixelatedPass;
exports.BloomPass = BloomPass.BloomPass;
exports.ARButton = ARButton.ARButton;
exports.OculusHandModel = OculusHandModel.OculusHandModel;
exports.OculusHandPointerModel = OculusHandPointerModel.OculusHandPointerModel;
exports.createText = Text2D.createText;
exports.VRButton = VRButton.VRButton;
exports.XRControllerModelFactory = XRControllerModelFactory.XRControllerModelFactory;
exports.XREstimatedLight = XREstimatedLight.XREstimatedLight;
exports.XRHandMeshModel = XRHandMeshModel.XRHandMeshModel;
exports.XRHandModelFactory = XRHandModelFactory.XRHandModelFactory;
exports.XRHandPrimitiveModel = XRHandPrimitiveModel.XRHandPrimitiveModel;
exports.ParametricGeometries = ParametricGeometries.ParametricGeometries;
exports.ParametricGeometry = ParametricGeometry.ParametricGeometry;
exports.ConvexGeometry = ConvexGeometry.ConvexGeometry;
exports.LightningStrike = LightningStrike.LightningStrike;
exports.RoundedBoxGeometry = RoundedBoxGeometry.RoundedBoxGeometry;
exports.BoxLineGeometry = BoxLineGeometry.BoxLineGeometry;
exports.DecalGeometry = DecalGeometry.DecalGeometry;
exports.DecalVertex = DecalGeometry.DecalVertex;
exports.TeapotGeometry = TeapotGeometry.TeapotGeometry;
exports.TextBufferGeometry = TextGeometry.TextGeometry;
exports.TextGeometry = TextGeometry.TextGeometry;
exports.CSM = CSM.CSM;
exports.CSMFrustum = CSMFrustum.CSMFrustum;
exports.CSMHelper = CSMHelper.CSMHelper;
exports.CSMShader = CSMShader.CSMShader;
exports.ACESFilmicToneMappingShader = ACESFilmicToneMappingShader.ACESFilmicToneMappingShader;
exports.AfterimageShader = AfterimageShader.AfterimageShader;
exports.BasicShader = BasicShader.BasicShader;
exports.BleachBypassShader = BleachBypassShader.BleachBypassShader;
exports.BlendShader = BlendShader.BlendShader;
exports.BokehShader = BokehShader.BokehShader;
exports.BokehShader2 = BokehShader2.BokehShader2;
exports.BrightnessContrastShader = BrightnessContrastShader.BrightnessContrastShader;
exports.ColorCorrectionShader = ColorCorrectionShader.ColorCorrectionShader;
exports.ColorifyShader = ColorifyShader.ColorifyShader;
exports.ConvolutionShader = ConvolutionShader.ConvolutionShader;
exports.CopyShader = CopyShader.CopyShader;
exports.DOFMipMapShader = DOFMipMapShader.DOFMipMapShader;
exports.BlurShaderUtils = DepthLimitedBlurShader.BlurShaderUtils;
exports.DepthLimitedBlurShader = DepthLimitedBlurShader.DepthLimitedBlurShader;
exports.DigitalGlitch = DigitalGlitch.DigitalGlitch;
exports.DotScreenShader = DotScreenShader.DotScreenShader;
exports.FXAAShader = FXAAShader.FXAAShader;
exports.FilmShader = FilmShader.FilmShader;
exports.FocusShader = FocusShader.FocusShader;
exports.FreiChenShader = FreiChenShader.FreiChenShader;
exports.FresnelShader = FresnelShader.FresnelShader;
exports.GammaCorrectionShader = GammaCorrectionShader.GammaCorrectionShader;
exports.GodRaysCombineShader = GodRaysShader.GodRaysCombineShader;
exports.GodRaysDepthMaskShader = GodRaysShader.GodRaysDepthMaskShader;
exports.GodRaysFakeSunShader = GodRaysShader.GodRaysFakeSunShader;
exports.GodRaysGenerateShader = GodRaysShader.GodRaysGenerateShader;
exports.HalftoneShader = HalftoneShader.HalftoneShader;
exports.HorizontalBlurShader = HorizontalBlurShader.HorizontalBlurShader;
exports.HorizontalTiltShiftShader = HorizontalTiltShiftShader.HorizontalTiltShiftShader;
exports.HueSaturationShader = HueSaturationShader.HueSaturationShader;
exports.KaleidoShader = KaleidoShader.KaleidoShader;
exports.LuminosityHighPassShader = LuminosityHighPassShader.LuminosityHighPassShader;
exports.LuminosityShader = LuminosityShader.LuminosityShader;
exports.MirrorShader = MirrorShader.MirrorShader;
exports.NormalMapShader = NormalMapShader.NormalMapShader;
exports.ParallaxShader = ParallaxShader.ParallaxShader;
exports.PixelShader = PixelShader.PixelShader;
exports.RGBShiftShader = RGBShiftShader.RGBShiftShader;
exports.SAOShader = SAOShader.SAOShader;
exports.SMAABlendShader = SMAAShader.SMAABlendShader;
exports.SMAAEdgesShader = SMAAShader.SMAAEdgesShader;
exports.SMAAWeightsShader = SMAAShader.SMAAWeightsShader;
exports.SSAOBlurShader = SSAOShader.SSAOBlurShader;
exports.SSAODepthShader = SSAOShader.SSAODepthShader;
exports.SSAOShader = SSAOShader.SSAOShader;
exports.SSRBlurShader = SSRShader.SSRBlurShader;
exports.SSRDepthShader = SSRShader.SSRDepthShader;
exports.SSRShader = SSRShader.SSRShader;
exports.SepiaShader = SepiaShader.SepiaShader;
exports.SobelOperatorShader = SobelOperatorShader.SobelOperatorShader;
exports.SubsurfaceScatteringShader = SubsurfaceScatteringShader.SubsurfaceScatteringShader;
exports.TechnicolorShader = TechnicolorShader.TechnicolorShader;
exports.ToneMapShader = ToneMapShader.ToneMapShader;
exports.ToonShader1 = ToonShader.ToonShader1;
exports.ToonShader2 = ToonShader.ToonShader2;
exports.ToonShaderDotted = ToonShader.ToonShaderDotted;
exports.ToonShaderHatching = ToonShader.ToonShaderHatching;
exports.TriangleBlurShader = TriangleBlurShader.TriangleBlurShader;
exports.UnpackDepthRGBAShader = UnpackDepthRGBAShader.UnpackDepthRGBAShader;
exports.VerticalBlurShader = VerticalBlurShader.VerticalBlurShader;
exports.VerticalTiltShiftShader = VerticalTiltShiftShader.VerticalTiltShiftShader;
exports.VignetteShader = VignetteShader.VignetteShader;
exports.VolumeRenderShader1 = VolumeShader.VolumeRenderShader1;
exports.WaterRefractionShader = WaterRefractionShader.WaterRefractionShader;
exports.SelectionHelper = SelectionHelper.SelectionHelper;
exports.SelectionBox = SelectionBox.SelectionBox;
exports.AmmoPhysics = AmmoPhysics.AmmoPhysics;
exports.ParallaxBarrierEffect = ParallaxBarrierEffect.ParallaxBarrierEffect;
exports.PeppersGhostEffect = PeppersGhostEffect.PeppersGhostEffect;
exports.OutlineEffect = OutlineEffect.OutlineEffect;
exports.AnaglyphEffect = AnaglyphEffect.AnaglyphEffect;
exports.AsciiEffect = AsciiEffect.AsciiEffect;
exports.StereoEffect = StereoEffect.StereoEffect;
exports.FBXLoader = FBXLoader.FBXLoader;
exports.Font = FontLoader.Font;
exports.FontLoader = FontLoader.FontLoader;
exports.TGALoader = TGALoader.TGALoader;
exports.LUTCubeLoader = LUTCubeLoader.LUTCubeLoader;
exports.NRRDLoader = NRRDLoader.NRRDLoader;
exports.STLLoader = STLLoader.STLLoader;
exports.MTLLoader = MTLLoader.MTLLoader;
exports.XLoader = XLoader.XLoader;
exports.BVHLoader = BVHLoader.BVHLoader;
exports.KMZLoader = KMZLoader.KMZLoader;
exports.VRMLoader = VRMLoader.VRMLoader;
exports.VRMLLoader = VRMLLoader.VRMLLoader;
exports.KTX2Loader = KTX2Loader.KTX2Loader;
exports.LottieLoader = LottieLoader.LottieLoader;
exports.TTFLoader = TTFLoader.TTFLoader;
exports.RGBELoader = RGBELoader.RGBELoader;
exports.AssimpLoader = AssimpLoader.AssimpLoader;
exports.ColladaLoader = ColladaLoader.ColladaLoader;
exports.MDDLoader = MDDLoader.MDDLoader;
exports.EXRLoader = EXRLoader.EXRLoader;
exports.ThreeMFLoader = _3MFLoader.ThreeMFLoader;
exports.XYZLoader = XYZLoader.XYZLoader;
exports.VTKLoader = VTKLoader.VTKLoader;
exports.LUT3dlLoader = LUT3dlLoader.LUT3dlLoader;
exports.DDSLoader = DDSLoader.DDSLoader;
exports.PVRLoader = PVRLoader.PVRLoader;
exports.GCodeLoader = GCodeLoader.GCodeLoader;
exports.BasisTextureLoader = BasisTextureLoader.BasisTextureLoader;
exports.TDSLoader = TDSLoader.TDSLoader;
exports.LDrawLoader = LDrawLoader.LDrawLoader;
exports.GLTFLoader = GLTFLoader.GLTFLoader;
exports.LWO3Parser = LWO3Parser.LWO3Parser;
exports.LWO2Parser = LWO2Parser.LWO2Parser;
exports.IFFParser = IFFParser.IFFParser;
exports.SVGLoader = SVGLoader.SVGLoader;
exports.Rhino3dmLoader = _3DMLoader.Rhino3dmLoader;
exports.OBJLoader = OBJLoader.OBJLoader;
exports.AMFLoader = AMFLoader.AMFLoader;
exports.MMDLoader = MMDLoader.MMDLoader;
exports.MD2Loader = MD2Loader.MD2Loader;
exports.KTXLoader = KTXLoader.KTXLoader;
exports.TiltLoader = TiltLoader.TiltLoader;
exports.DRACOLoader = DRACOLoader.DRACOLoader;
exports.HDRCubeTextureLoader = HDRCubeTextureLoader.HDRCubeTextureLoader;
exports.PDBLoader = PDBLoader.PDBLoader;
exports.PRWMLoader = PRWMLoader.PRWMLoader;
exports.RGBMLoader = RGBMLoader.RGBMLoader;
exports.VOXDataTexture3D = VOXLoader.VOXDataTexture3D;
exports.VOXLoader = VOXLoader.VOXLoader;
exports.VOXMesh = VOXLoader.VOXMesh;
exports.NodeMaterialLoaderUtils = NodeMaterialLoader.NodeMaterialLoaderUtils;
exports.PCDLoader = PCDLoader.PCDLoader;
exports.LWOLoader = LWOLoader.LWOLoader;
exports.PLYLoader = PLYLoader.PLYLoader;
exports.LineSegmentsGeometry = LineSegmentsGeometry.LineSegmentsGeometry;
exports.LineGeometry = LineGeometry.LineGeometry;
exports.Wireframe = Wireframe.Wireframe;
exports.WireframeGeometry2 = WireframeGeometry2.WireframeGeometry2;
exports.Line2 = Line2.Line2;
exports.LineMaterial = LineMaterial.LineMaterial;
exports.LineSegments2 = LineSegments2.LineSegments2;
exports.LightProbeHelper = LightProbeHelper.LightProbeHelper;
exports.VertexTangentsHelper = VertexTangentsHelper.VertexTangentsHelper;
exports.PositionalAudioHelper = PositionalAudioHelper.PositionalAudioHelper;
exports.VertexNormalsHelper = VertexNormalsHelper.VertexNormalsHelper;
exports.RectAreaLightHelper = RectAreaLightHelper.RectAreaLightHelper;
exports.RectAreaLightUniformsLib = RectAreaLightUniformsLib.RectAreaLightUniformsLib;
exports.LightProbeGenerator = LightProbeGenerator.LightProbeGenerator;
exports.calcBSplineDerivatives = NURBSUtils.calcBSplineDerivatives;
exports.calcBSplinePoint = NURBSUtils.calcBSplinePoint;
exports.calcBasisFunctionDerivatives = NURBSUtils.calcBasisFunctionDerivatives;
exports.calcBasisFunctions = NURBSUtils.calcBasisFunctions;
exports.calcKoverI = NURBSUtils.calcKoverI;
exports.calcNURBSDerivatives = NURBSUtils.calcNURBSDerivatives;
exports.calcRationalCurveDerivatives = NURBSUtils.calcRationalCurveDerivatives;
exports.calcSurfacePoint = NURBSUtils.calcSurfacePoint;
exports.findSpan = NURBSUtils.findSpan;
exports.NURBSCurve = NURBSCurve.NURBSCurve;
exports.NURBSSurface = NURBSSurface.NURBSSurface;
exports.CinquefoilKnot = CurveExtras.CinquefoilKnot;
exports.DecoratedTorusKnot4a = CurveExtras.DecoratedTorusKnot4a;
exports.DecoratedTorusKnot4b = CurveExtras.DecoratedTorusKnot4b;
exports.DecoratedTorusKnot5a = CurveExtras.DecoratedTorusKnot5a;
exports.DecoratedTorusKnot5c = CurveExtras.DecoratedTorusKnot5c;
exports.FigureEightPolynomialKnot = CurveExtras.FigureEightPolynomialKnot;
exports.GrannyKnot = CurveExtras.GrannyKnot;
exports.HeartCurve = CurveExtras.HeartCurve;
exports.HelixCurve = CurveExtras.HelixCurve;
exports.KnotCurve = CurveExtras.KnotCurve;
exports.TorusKnot = CurveExtras.TorusKnot;
exports.TrefoilKnot = CurveExtras.TrefoilKnot;
exports.TrefoilPolynomialKnot = CurveExtras.TrefoilPolynomialKnot;
exports.VivianiCurve = CurveExtras.VivianiCurve;
exports.scaleTo = CurveExtras.scaleTo;
exports.Face3 = Geometry.Face3;
exports.Geometry = Geometry.Geometry;
exports.MeshoptDecoder = MeshoptDecoder.MeshoptDecoder;
exports.MotionController = MotionControllers.MotionController;
exports.MotionControllerConstants = MotionControllers.MotionControllerConstants;
exports.fetchProfile = MotionControllers.fetchProfile;
exports.fetchProfilesList = MotionControllers.fetchProfilesList;
exports.BRDF_GGX = BSDFs.BRDF_GGX;
exports.BRDF_Lambert = BSDFs.BRDF_Lambert;
exports.D_GGX = BSDFs.D_GGX;
exports.F_Schlick = BSDFs.F_Schlick;
exports.PhysicalLightingModel = BSDFs.PhysicalLightingModel;
exports.RE_Direct_Physical = BSDFs.RE_Direct_Physical;
exports.V_GGX_SmithCorrelated = BSDFs.V_GGX_SmithCorrelated;
exports.getDistanceAttenuation = BSDFs.getDistanceAttenuation;
exports.LineBasicNodeMaterial = LineBasicNodeMaterial;
exports.MeshBasicNodeMaterial = MeshBasicNodeMaterial;
exports.MeshStandardNodeMaterial = MeshStandardNodeMaterial;
exports.PointsNodeMaterial = PointsNodeMaterial;
exports.EPSILON = ShaderNode.EPSILON;
exports.INFINITY = ShaderNode.INFINITY;
exports.ShaderNode = ShaderNode.ShaderNode;
exports.abs = ShaderNode.abs;
exports.acos = ShaderNode.acos;
exports.add = ShaderNode.add;
exports.addTo = ShaderNode.addTo;
exports.alphaTest = ShaderNode.alphaTest;
exports.and = ShaderNode.and;
exports.asin = ShaderNode.asin;
exports.assign = ShaderNode.assign;
exports.atan = ShaderNode.atan;
exports.attribute = ShaderNode.attribute;
exports.bitAnd = ShaderNode.bitAnd;
exports.bitOr = ShaderNode.bitOr;
exports.bitXor = ShaderNode.bitXor;
exports.bmat3 = ShaderNode.bmat3;
exports.bmat4 = ShaderNode.bmat4;
exports.bool = ShaderNode.bool;
exports.buffer = ShaderNode.buffer;
exports.bvec2 = ShaderNode.bvec2;
exports.bvec3 = ShaderNode.bvec3;
exports.bvec4 = ShaderNode.bvec4;
exports.cameraPosition = ShaderNode.cameraPosition;
exports.ceil = ShaderNode.ceil;
exports.clamp = ShaderNode.clamp;
exports.color = ShaderNode.color;
exports.cond = ShaderNode.cond;
exports.cos = ShaderNode.cos;
exports.cross = ShaderNode.cross;
exports.dFdx = ShaderNode.dFdx;
exports.dFdy = ShaderNode.dFdy;
exports.degrees = ShaderNode.degrees;
exports.diffuseColor = ShaderNode.diffuseColor;
exports.distance = ShaderNode.distance;
exports.div = ShaderNode.div;
exports.dot = ShaderNode.dot;
exports.element = ShaderNode.element;
exports.equal = ShaderNode.equal;
exports.exp = ShaderNode.exp;
exports.exp2 = ShaderNode.exp2;
exports.faceforward = ShaderNode.faceforward;
exports.float = ShaderNode.float;
exports.floor = ShaderNode.floor;
exports.fract = ShaderNode.fract;
exports.greaterThan = ShaderNode.greaterThan;
exports.greaterThanEqual = ShaderNode.greaterThanEqual;
exports.imat3 = ShaderNode.imat3;
exports.imat4 = ShaderNode.imat4;
exports.int = ShaderNode.int;
exports.inversesqrt = ShaderNode.inversesqrt;
exports.invert = ShaderNode.invert;
exports.ivec2 = ShaderNode.ivec2;
exports.ivec3 = ShaderNode.ivec3;
exports.ivec4 = ShaderNode.ivec4;
exports.join = ShaderNode.join;
exports.label = ShaderNode.label;
exports.length = ShaderNode.length;
exports.lessThan = ShaderNode.lessThan;
exports.lessThanEqual = ShaderNode.lessThanEqual;
exports.log = ShaderNode.log;
exports.log2 = ShaderNode.log2;
exports.mat3 = ShaderNode.mat3;
exports.mat4 = ShaderNode.mat4;
exports.max = ShaderNode.max;
exports.metalness = ShaderNode.metalness;
exports.min = ShaderNode.min;
exports.mix = ShaderNode.mix;
exports.mod = ShaderNode.mod;
exports.mul = ShaderNode.mul;
exports.negate = ShaderNode.negate;
exports.nodeObject = ShaderNode.nodeObject;
exports.normalGeometry = ShaderNode.normalGeometry;
exports.normalLocal = ShaderNode.normalLocal;
exports.normalView = ShaderNode.normalView;
exports.normalWorld = ShaderNode.normalWorld;
exports.normalize = ShaderNode.normalize;
exports.or = ShaderNode.or;
exports.positionLocal = ShaderNode.positionLocal;
exports.positionView = ShaderNode.positionView;
exports.positionViewDirection = ShaderNode.positionViewDirection;
exports.positionWorld = ShaderNode.positionWorld;
exports.pow = ShaderNode.pow;
exports.pow2 = ShaderNode.pow2;
exports.pow3 = ShaderNode.pow3;
exports.pow4 = ShaderNode.pow4;
exports.radians = ShaderNode.radians;
exports.reflect = ShaderNode.reflect;
exports.refract = ShaderNode.refract;
exports.remainder = ShaderNode.remainder;
exports.roughness = ShaderNode.roughness;
exports.round = ShaderNode.round;
exports.sampler = ShaderNode.sampler;
exports.saturate = ShaderNode.saturate;
exports.shiftLeft = ShaderNode.shiftLeft;
exports.shiftRight = ShaderNode.shiftRight;
exports.sign = ShaderNode.sign;
exports.sin = ShaderNode.sin;
exports.smoothstep = ShaderNode.smoothstep;
exports.specularColor = ShaderNode.specularColor;
exports.sqrt = ShaderNode.sqrt;
exports.step = ShaderNode.step;
exports.sub = ShaderNode.sub;
exports.tan = ShaderNode.tan;
exports.temp = ShaderNode.temp;
exports.texture = ShaderNode.texture;
exports.transformDirection = ShaderNode.transformDirection;
exports.transformedNormalView = ShaderNode.transformedNormalView;
exports.uint = ShaderNode.uint;
exports.umat3 = ShaderNode.umat3;
exports.umat4 = ShaderNode.umat4;
exports.uniform = ShaderNode.uniform;
exports.uv = ShaderNode.uv;
exports.uvec2 = ShaderNode.uvec2;
exports.uvec3 = ShaderNode.uvec3;
exports.uvec4 = ShaderNode.uvec4;
exports.vec2 = ShaderNode.vec2;
exports.vec3 = ShaderNode.vec3;
exports.vec4 = ShaderNode.vec4;
exports.viewMatrix = ShaderNode.viewMatrix;
exports.xor = ShaderNode.xor;
exports.ArrayUniformNode = ArrayUniformNode;
exports.AttributeNode = AttributeNode;
exports.BypassNode = BypassNode;
exports.CodeNode = CodeNode;
exports.ContextNode = ContextNode;
exports.ConstNode = ConstNode;
exports.ExpressionNode = ExpressionNode;
exports.FunctionCallNode = FunctionCallNode;
exports.FunctionNode = FunctionNode;
exports.Node = Node;
exports.NodeAttribute = NodeAttribute;
exports.NodeBuilder = NodeBuilder.default;
exports.NodeCode = NodeCode;
exports.NodeFrame = NodeFrame;
exports.NodeFunctionInput = NodeFunctionInput;
exports.NodeKeywords = NodeKeywords;
exports.NodeUniform = NodeUniform;
exports.NodeVar = NodeVar;
exports.NodeVary = NodeVary;
exports.PropertyNode = PropertyNode;
exports.TempNode = TempNode;
exports.UniformNode = UniformNode;
exports.VarNode = VarNode;
exports.VaryNode = VaryNode;
exports.BufferNode = BufferNode;
exports.CameraNode = CameraNode;
exports.CubeTextureNode = CubeTextureNode;
exports.MaterialNode = MaterialNode;
exports.MaterialReferenceNode = MaterialReferenceNode;
exports.ModelNode = ModelNode;
exports.ModelViewProjectionNode = ModelViewProjectionNode;
exports.NormalNode = NormalNode;
exports.Object3DNode = Object3DNode;
exports.PointUVNode = PointUVNode;
exports.PositionNode = PositionNode;
exports.ReferenceNode = ReferenceNode;
exports.ReflectNode = ReflectNode;
exports.SkinningNode = SkinningNode;
exports.TextureNode = TextureNode;
exports.UVNode = UVNode;
exports.ColorSpaceNode = ColorSpaceNode.default;
exports.NormalMapNode = NormalMapNode;
exports.MathNode = MathNode;
exports.OperatorNode = OperatorNode;
exports.CondNode = CondNode;
exports.LightContextNode = LightContextNode;
exports.LightNode = LightNode;
exports.LightsNode = LightsNode;
exports.ArrayElementNode = ArrayElementNode;
exports.ConvertNode = ConvertNode;
exports.JoinNode = JoinNode;
exports.SplitNode = SplitNode;
exports.SpriteSheetUVNode = SpriteSheetUVNode;
exports.MatcapUVNode = MatcapUVNode;
exports.OscNode = OscNode;
exports.TimerNode = TimerNode;
exports.CheckerNode = CheckerNode;
exports.FogNode = FogNode;
exports.FogRangeNode = FogRangeNode;
exports.NodeLoader = NodeLoader;
exports.NodeObjectLoader = NodeObjectLoader;
