// three-stdlib/renderers/webgpu/nodes/WebGPUNodes.js
"use strict";
const WebGPUNodeBuilder = require("./WebGPUNodeBuilder.js");
const NodeFrame = require("../../nodes/core/NodeFrame.js");
class WebGPUNodes {
  constructor(renderer) {
    this.renderer = renderer;
    this.nodeFrame = new NodeFrame();
    this.builders = /* @__PURE__ */ new WeakMap();
  }
  get(material) {
    let nodeBuilder = this.builders.get(material);
    if (nodeBuilder === void 0) {
      nodeBuilder = new WebGPUNodeBuilder(material, this.renderer).build();
      this.builders.set(material, nodeBuilder);
    }
    return nodeBuilder;
  }
  remove(material) {
    this.builders.delete(material);
  }
  updateFrame() {
    this.nodeFrame.update();
  }
  update(object, camera) {
    const material = object.material;
    const nodeBuilder = this.get(material);
    const nodeFrame = this.nodeFrame;
    nodeFrame.material = material;
    nodeFrame.camera = camera;
    nodeFrame.object = object;
    for (let node of nodeBuilder.updateNodes) {
      nodeFrame.updateNode(node);
    }
  }
  dispose() {
    this.builders = /* @__PURE__ */ new WeakMap();
  }
}
module.exports = WebGPUNodes;
