// three-stdlib/postprocessing/DotScreenPass.js
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const Pass = require("./Pass.js");
const THREE = require("three");
const DotScreenShader = require("../shaders/DotScreenShader.js");
class DotScreenPass extends Pass.Pass {
  constructor(center, angle, scale) {
    super();
    if (DotScreenShader.DotScreenShader === void 0)
      console.error("THREE.DotScreenPass relies on THREE.DotScreenShader");
    const shader = DotScreenShader.DotScreenShader;
    this.uniforms = THREE.UniformsUtils.clone(shader.uniforms);
    if (center !== void 0)
      this.uniforms["center"].value.copy(center);
    if (angle !== void 0)
      this.uniforms["angle"].value = angle;
    if (scale !== void 0)
      this.uniforms["scale"].value = scale;
    this.material = new THREE.ShaderMaterial({
      uniforms: this.uniforms,
      vertexShader: shader.vertexShader,
      fragmentShader: shader.fragmentShader
    });
    this.fsQuad = new Pass.FullScreenQuad(this.material);
  }
  render(renderer, writeBuffer, readBuffer) {
    this.uniforms["tDiffuse"].value = readBuffer.texture;
    this.uniforms["tSize"].value.set(readBuffer.width, readBuffer.height);
    if (this.renderToScreen) {
      renderer.setRenderTarget(null);
      this.fsQuad.render(renderer);
    } else {
      renderer.setRenderTarget(writeBuffer);
      if (this.clear)
        renderer.clear();
      this.fsQuad.render(renderer);
    }
  }
}
exports.DotScreenPass = DotScreenPass;
