// three-stdlib/postprocessing/DotScreenPass.mjs
import { Pass, FullScreenQuad } from "./Pass.mjs";
import { UniformsUtils, ShaderMaterial } from "three";
import { DotScreenShader } from "../shaders/DotScreenShader.mjs";
class DotScreenPass extends Pass {
  constructor(center, angle, scale) {
    super();
    if (DotScreenShader === void 0)
      console.error("THREE.DotScreenPass relies on THREE.DotScreenShader");
    const shader = DotScreenShader;
    this.uniforms = UniformsUtils.clone(shader.uniforms);
    if (center !== void 0)
      this.uniforms["center"].value.copy(center);
    if (angle !== void 0)
      this.uniforms["angle"].value = angle;
    if (scale !== void 0)
      this.uniforms["scale"].value = scale;
    this.material = new ShaderMaterial({
      uniforms: this.uniforms,
      vertexShader: shader.vertexShader,
      fragmentShader: shader.fragmentShader
    });
    this.fsQuad = new FullScreenQuad(this.material);
  }
  render(renderer, writeBuffer, readBuffer) {
    this.uniforms["tDiffuse"].value = readBuffer.texture;
    this.uniforms["tSize"].value.set(readBuffer.width, readBuffer.height);
    if (this.renderToScreen) {
      renderer.setRenderTarget(null);
      this.fsQuad.render(renderer);
    } else {
      renderer.setRenderTarget(writeBuffer);
      if (this.clear)
        renderer.clear();
      this.fsQuad.render(renderer);
    }
  }
}
export {
  DotScreenPass
};
