// three-stdlib/nodes/Nodes.mjs
import ArrayUniformNode from "./core/ArrayUniformNode.mjs";
import { default as default2 } from "./core/ArrayUniformNode.mjs";
import AttributeNode from "./core/AttributeNode.mjs";
import { default as default3 } from "./core/AttributeNode.mjs";
import BypassNode from "./core/BypassNode.mjs";
import { default as default4 } from "./core/BypassNode.mjs";
import CodeNode from "./core/CodeNode.mjs";
import { default as default5 } from "./core/CodeNode.mjs";
import ConstNode from "./core/ConstNode.mjs";
import { default as default6 } from "./core/ConstNode.mjs";
import ContextNode from "./core/ContextNode.mjs";
import { default as default7 } from "./core/ContextNode.mjs";
import ExpressionNode from "./core/ExpressionNode.mjs";
import { default as default8 } from "./core/ExpressionNode.mjs";
import FunctionCallNode from "./core/FunctionCallNode.mjs";
import { default as default9 } from "./core/FunctionCallNode.mjs";
import FunctionNode from "./core/FunctionNode.mjs";
import { default as default10 } from "./core/FunctionNode.mjs";
import Node from "./core/Node.mjs";
import { default as default11 } from "./core/Node.mjs";
import NodeAttribute from "./core/NodeAttribute.mjs";
import { default as default12 } from "./core/NodeAttribute.mjs";
import NodeBuilder from "./core/NodeBuilder.mjs";
import { default as default13 } from "./core/NodeBuilder.mjs";
import NodeCode from "./core/NodeCode.mjs";
import { default as default14 } from "./core/NodeCode.mjs";
import NodeFrame from "./core/NodeFrame.mjs";
import { default as default15 } from "./core/NodeFrame.mjs";
import NodeFunctionInput from "./core/NodeFunctionInput.mjs";
import { default as default16 } from "./core/NodeFunctionInput.mjs";
import NodeKeywords from "./core/NodeKeywords.mjs";
import { default as default17 } from "./core/NodeKeywords.mjs";
import NodeUniform from "./core/NodeUniform.mjs";
import { default as default18 } from "./core/NodeUniform.mjs";
import NodeVar from "./core/NodeVar.mjs";
import { default as default19 } from "./core/NodeVar.mjs";
import NodeVary from "./core/NodeVary.mjs";
import { default as default20 } from "./core/NodeVary.mjs";
import PropertyNode from "./core/PropertyNode.mjs";
import { default as default21 } from "./core/PropertyNode.mjs";
import TempNode from "./core/TempNode.mjs";
import { default as default22 } from "./core/TempNode.mjs";
import UniformNode from "./core/UniformNode.mjs";
import { default as default23 } from "./core/UniformNode.mjs";
import VarNode from "./core/VarNode.mjs";
import { default as default24 } from "./core/VarNode.mjs";
import VaryNode from "./core/VaryNode.mjs";
import { default as default25 } from "./core/VaryNode.mjs";
import BufferNode from "./accessors/BufferNode.mjs";
import { default as default26 } from "./accessors/BufferNode.mjs";
import CameraNode from "./accessors/CameraNode.mjs";
import { default as default27 } from "./accessors/CameraNode.mjs";
import CubeTextureNode from "./accessors/CubeTextureNode.mjs";
import { default as default28 } from "./accessors/CubeTextureNode.mjs";
import MaterialNode from "./accessors/MaterialNode.mjs";
import { default as default29 } from "./accessors/MaterialNode.mjs";
import MaterialReferenceNode from "./accessors/MaterialReferenceNode.mjs";
import { default as default30 } from "./accessors/MaterialReferenceNode.mjs";
import ModelNode from "./accessors/ModelNode.mjs";
import { default as default31 } from "./accessors/ModelNode.mjs";
import ModelViewProjectionNode from "./accessors/ModelViewProjectionNode.mjs";
import { default as default32 } from "./accessors/ModelViewProjectionNode.mjs";
import NormalNode from "./accessors/NormalNode.mjs";
import { default as default33 } from "./accessors/NormalNode.mjs";
import Object3DNode from "./accessors/Object3DNode.mjs";
import { default as default34 } from "./accessors/Object3DNode.mjs";
import PointUVNode from "./accessors/PointUVNode.mjs";
import { default as default35 } from "./accessors/PointUVNode.mjs";
import PositionNode from "./accessors/PositionNode.mjs";
import { default as default36 } from "./accessors/PositionNode.mjs";
import ReferenceNode from "./accessors/ReferenceNode.mjs";
import { default as default37 } from "./accessors/ReferenceNode.mjs";
import ReflectNode from "./accessors/ReflectNode.mjs";
import { default as default38 } from "./accessors/ReflectNode.mjs";
import SkinningNode from "./accessors/SkinningNode.mjs";
import { default as default39 } from "./accessors/SkinningNode.mjs";
import TextureNode from "./accessors/TextureNode.mjs";
import { default as default40 } from "./accessors/TextureNode.mjs";
import UVNode from "./accessors/UVNode.mjs";
import { default as default41 } from "./accessors/UVNode.mjs";
import ColorSpaceNode from "./display/ColorSpaceNode.mjs";
import { default as default42 } from "./display/ColorSpaceNode.mjs";
import NormalMapNode from "./display/NormalMapNode.mjs";
import { default as default43 } from "./display/NormalMapNode.mjs";
import MathNode from "./math/MathNode.mjs";
import { default as default44 } from "./math/MathNode.mjs";
import OperatorNode from "./math/OperatorNode.mjs";
import { default as default45 } from "./math/OperatorNode.mjs";
import CondNode from "./math/CondNode.mjs";
import { default as default46 } from "./math/CondNode.mjs";
import LightContextNode from "./lights/LightContextNode.mjs";
import { default as default47 } from "./lights/LightContextNode.mjs";
import LightNode from "./lights/LightNode.mjs";
import { default as default48 } from "./lights/LightNode.mjs";
import LightsNode from "./lights/LightsNode.mjs";
import { default as default49 } from "./lights/LightsNode.mjs";
import ArrayElementNode from "./utils/ArrayElementNode.mjs";
import { default as default50 } from "./utils/ArrayElementNode.mjs";
import ConvertNode from "./utils/ConvertNode.mjs";
import { default as default51 } from "./utils/ConvertNode.mjs";
import JoinNode from "./utils/JoinNode.mjs";
import { default as default52 } from "./utils/JoinNode.mjs";
import SplitNode from "./utils/SplitNode.mjs";
import { default as default53 } from "./utils/SplitNode.mjs";
import SpriteSheetUVNode from "./utils/SpriteSheetUVNode.mjs";
import { default as default54 } from "./utils/SpriteSheetUVNode.mjs";
import MatcapUVNode from "./utils/MatcapUVNode.mjs";
import { default as default55 } from "./utils/MatcapUVNode.mjs";
import OscNode from "./utils/OscNode.mjs";
import { default as default56 } from "./utils/OscNode.mjs";
import TimerNode from "./utils/TimerNode.mjs";
import { default as default57 } from "./utils/TimerNode.mjs";
import NodeLoader from "./loaders/NodeLoader.mjs";
import { default as default58 } from "./loaders/NodeLoader.mjs";
import NodeObjectLoader from "./loaders/NodeObjectLoader.mjs";
import { default as default59 } from "./loaders/NodeObjectLoader.mjs";
import NodeMaterialLoader from "./loaders/NodeMaterialLoader.mjs";
import { default as default60 } from "./loaders/NodeMaterialLoader.mjs";
import CheckerNode from "./procedural/CheckerNode.mjs";
import { default as default61 } from "./procedural/CheckerNode.mjs";
import FogNode from "./fog/FogNode.mjs";
import { default as default62 } from "./fog/FogNode.mjs";
import FogRangeNode from "./fog/FogRangeNode.mjs";
import { default as default63 } from "./fog/FogRangeNode.mjs";
import { NodeShaderStage, NodeType, NodeUpdateType } from "./core/constants.mjs";
import { BRDF_GGX, BRDF_Lambert, D_GGX, F_Schlick, PhysicalLightingModel, RE_Direct_Physical, V_GGX_SmithCorrelated, getDistanceAttenuation } from "./functions/BSDFs.mjs";
import "./materials/Materials.mjs";
import { EPSILON, INFINITY, ShaderNode, abs, acos, add, addTo, alphaTest, and, asin, assign, atan, attribute, bitAnd, bitOr, bitXor, bmat3, bmat4, bool, buffer, bvec2, bvec3, bvec4, cameraPosition, ceil, clamp, color, cond, cos, cross, dFdx, dFdy, degrees, diffuseColor, distance, div, dot, element, equal, exp, exp2, faceforward, float, floor, fract, greaterThan, greaterThanEqual, imat3, imat4, int, inversesqrt, invert, ivec2, ivec3, ivec4, join, label, length, lessThan, lessThanEqual, log, log2, mat3, mat4, max, metalness, min, mix, mod, mul, negate, nodeObject, normalGeometry, normalLocal, normalView, normalWorld, normalize, or, positionLocal, positionView, positionViewDirection, positionWorld, pow, pow2, pow3, pow4, radians, reflect, refract, remainder, roughness, round, sampler, saturate, shiftLeft, shiftRight, sign, sin, smoothstep, specularColor, sqrt, step, sub, tan, temp, texture, transformDirection, transformedNormalView, uint, umat3, umat4, uniform, uv, uvec2, uvec3, uvec4, vec2, vec3, vec4, viewMatrix, xor } from "./ShaderNode.mjs";
import { default as default64 } from "./materials/LineBasicNodeMaterial.mjs";
import { default as default65 } from "./materials/MeshBasicNodeMaterial.mjs";
import { default as default66 } from "./materials/MeshStandardNodeMaterial.mjs";
import { default as default67 } from "./materials/PointsNodeMaterial.mjs";
const nodeLib = {
  ArrayUniformNode,
  AttributeNode,
  BypassNode,
  CodeNode,
  ContextNode,
  ConstNode,
  ExpressionNode,
  FunctionCallNode,
  FunctionNode,
  Node,
  NodeAttribute,
  NodeBuilder,
  NodeCode,
  NodeFrame,
  NodeFunctionInput,
  NodeKeywords,
  NodeUniform,
  NodeVar,
  NodeVary,
  PropertyNode,
  TempNode,
  UniformNode,
  VarNode,
  VaryNode,
  BufferNode,
  CameraNode,
  CubeTextureNode,
  MaterialNode,
  MaterialReferenceNode,
  ModelNode,
  ModelViewProjectionNode,
  NormalNode,
  Object3DNode,
  PointUVNode,
  PositionNode,
  ReferenceNode,
  ReflectNode,
  SkinningNode,
  TextureNode,
  UVNode,
  ColorSpaceNode,
  NormalMapNode,
  MathNode,
  OperatorNode,
  CondNode,
  LightContextNode,
  LightNode,
  LightsNode,
  ArrayElementNode,
  ConvertNode,
  JoinNode,
  SplitNode,
  SpriteSheetUVNode,
  MatcapUVNode,
  OscNode,
  TimerNode,
  CheckerNode,
  FogNode,
  FogRangeNode,
  NodeLoader,
  NodeObjectLoader,
  NodeMaterialLoader
};
const fromType = (type) => {
  return new nodeLib[type]();
};
export {
  default50 as ArrayElementNode,
  default2 as ArrayUniformNode,
  default3 as AttributeNode,
  BRDF_GGX,
  BRDF_Lambert,
  default26 as BufferNode,
  default4 as BypassNode,
  default27 as CameraNode,
  default61 as CheckerNode,
  default5 as CodeNode,
  default42 as ColorSpaceNode,
  default46 as CondNode,
  default6 as ConstNode,
  default7 as ContextNode,
  default51 as ConvertNode,
  default28 as CubeTextureNode,
  D_GGX,
  EPSILON,
  default8 as ExpressionNode,
  F_Schlick,
  default62 as FogNode,
  default63 as FogRangeNode,
  default9 as FunctionCallNode,
  default10 as FunctionNode,
  INFINITY,
  default52 as JoinNode,
  default47 as LightContextNode,
  default48 as LightNode,
  default49 as LightsNode,
  default64 as LineBasicNodeMaterial,
  default55 as MatcapUVNode,
  default29 as MaterialNode,
  default30 as MaterialReferenceNode,
  default44 as MathNode,
  default65 as MeshBasicNodeMaterial,
  default66 as MeshStandardNodeMaterial,
  default31 as ModelNode,
  default32 as ModelViewProjectionNode,
  default11 as Node,
  default12 as NodeAttribute,
  default13 as NodeBuilder,
  default14 as NodeCode,
  default15 as NodeFrame,
  default16 as NodeFunctionInput,
  default17 as NodeKeywords,
  default58 as NodeLoader,
  default60 as NodeMaterialLoader,
  default59 as NodeObjectLoader,
  NodeShaderStage,
  NodeType,
  default18 as NodeUniform,
  NodeUpdateType,
  default19 as NodeVar,
  default20 as NodeVary,
  default43 as NormalMapNode,
  default33 as NormalNode,
  default34 as Object3DNode,
  default45 as OperatorNode,
  default56 as OscNode,
  PhysicalLightingModel,
  default35 as PointUVNode,
  default67 as PointsNodeMaterial,
  default36 as PositionNode,
  default21 as PropertyNode,
  RE_Direct_Physical,
  default37 as ReferenceNode,
  default38 as ReflectNode,
  ShaderNode,
  default39 as SkinningNode,
  default53 as SplitNode,
  default54 as SpriteSheetUVNode,
  default22 as TempNode,
  default40 as TextureNode,
  default57 as TimerNode,
  default41 as UVNode,
  default23 as UniformNode,
  V_GGX_SmithCorrelated,
  default24 as VarNode,
  default25 as VaryNode,
  abs,
  acos,
  add,
  addTo,
  alphaTest,
  and,
  asin,
  assign,
  atan,
  attribute,
  bitAnd,
  bitOr,
  bitXor,
  bmat3,
  bmat4,
  bool,
  buffer,
  bvec2,
  bvec3,
  bvec4,
  cameraPosition,
  ceil,
  clamp,
  color,
  cond,
  cos,
  cross,
  dFdx,
  dFdy,
  degrees,
  diffuseColor,
  distance,
  div,
  dot,
  element,
  equal,
  exp,
  exp2,
  faceforward,
  float,
  floor,
  fract,
  fromType,
  getDistanceAttenuation,
  greaterThan,
  greaterThanEqual,
  imat3,
  imat4,
  int,
  inversesqrt,
  invert,
  ivec2,
  ivec3,
  ivec4,
  join,
  label,
  length,
  lessThan,
  lessThanEqual,
  log,
  log2,
  mat3,
  mat4,
  max,
  metalness,
  min,
  mix,
  mod,
  mul,
  negate,
  nodeObject,
  normalGeometry,
  normalLocal,
  normalView,
  normalWorld,
  normalize,
  or,
  positionLocal,
  positionView,
  positionViewDirection,
  positionWorld,
  pow,
  pow2,
  pow3,
  pow4,
  radians,
  reflect,
  refract,
  remainder,
  roughness,
  round,
  sampler,
  saturate,
  shiftLeft,
  shiftRight,
  sign,
  sin,
  smoothstep,
  specularColor,
  sqrt,
  step,
  sub,
  tan,
  temp,
  texture,
  transformDirection,
  transformedNormalView,
  uint,
  umat3,
  umat4,
  uniform,
  uv,
  uvec2,
  uvec3,
  uvec4,
  vec2,
  vec3,
  vec4,
  viewMatrix,
  xor
};
